{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Benchmarking notebook\n",
    "\n",
    "This notebook contains all of the methods and analysis for performing benchmarking.\n",
    "\n",
    "Each section contains a description to allow for more-easily walking through the code. We keep the benchmarking code in a jupyter-notebook as we antipate ourselves (and others) adding new parsers and using this for replicating/extending benchmarking."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Imports and dependencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# std library\n",
    "import os\n",
    "import glob\n",
    "from pathlib import Path\n",
    "from enum import Enum\n",
    "import numpy as np\n",
    "from dataclasses import dataclass\n",
    "from typing import List, Dict\n",
    "import pickle\n",
    "import json\n",
    "# non-std library\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Notebook parameters\n",
    "These parameters are used throughout the notebook for benchmarking. These parameters include paths, tuning, and other various parameter settings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_with_names = \"./abundance_new.tsv\"\n",
    "accession2taxid_file_path = \"../../database/krakenDB/taxonomy/nucl_gb.accession2taxid\"\n",
    "ncbi_map_file_path = \"ncbi_map.tab\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(ncbi_map_file_path):\n",
    "    !cat ../../database/krakenDB/taxonomy/nucl_gb.accession2taxid | grep \"NC_\" > ncbi_map.tab"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create mapping dictionary\n",
    "This step is used for creating a map from NCBI IDs to taxonomy IDs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_needed_ncbi_ids():\n",
    "    \"\"\"\n",
    "    **DEPRECATED FUNCTION**\n",
    "    was originally used to get NCBI ids from an output\n",
    "    fastviromexplorer file.\n",
    "    **DEPRECATED FUNCTION**\n",
    "    \"\"\"\n",
    "    file_with_names_opened = open(file_with_names)\n",
    "    \n",
    "    # grab all needed NCBI ids\n",
    "    ncbi_ids = []\n",
    "    line = file_with_names_opened.readline()\n",
    "    line_counter = 0\n",
    "    while(line):\n",
    "        if (line_counter > 0): # skip header\n",
    "            ncbi_id = line.split(\"\\t\")[0]\n",
    "            ncbi_ids.append(ncbi_id)\n",
    "        line = file_with_names_opened.readline()\n",
    "        line_counter += 1\n",
    "        \n",
    "    file_with_names_opened.close()\n",
    "    \n",
    "    return ncbi_ids\n",
    "\n",
    "ncbi_ids = get_needed_ncbi_ids()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_mapping_dictionary(map_file=ncbi_map_file_path):\n",
    "    \"\"\"\n",
    "    Creates dictionary for mapping NCBI ids to \n",
    "    taxonomy.\n",
    "    -------------------------\n",
    "    Notes:\n",
    "        1. cat nucl_gb.accession2taxid | grep \"NC_\" > ncbi_ids.tab\n",
    "        2. Uses ncbi_ids.tab\n",
    "    -------------------------\n",
    "    NC_033618\tNC_033618.1\t1931113\t1139918407\n",
    "    \"\"\"\n",
    "    # grab taxid mappings\n",
    "    acc2tax_open = open(map_file)\n",
    "    line = acc2tax_open.readline()\n",
    "    counter = 0\n",
    "    while(line):\n",
    "        line = line.split(\"\\t\")\n",
    "        accession = line[0]\n",
    "        taxid = line[2]\n",
    "        dictionary[accession] = taxid\n",
    "        counter += 1\n",
    "#         if (counter % 10000 == 0):\n",
    "#             print(f\"{counter} lines have been parsed \\n\")\n",
    "        line = acc2tax_open.readline()\n",
    "    acc2tax_open.close()\n",
    "    \n",
    "    return dictionary\n",
    "\n",
    "dictionary = create_mapping_dictionary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'2919554'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dictionary['NC_014459']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Parsers\n",
    "These methods are used for parsing all of the tools being benchmarked. These parsers are passed to a general parser function, giving a strategy-pattern-like method for obtaining result from tools."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parser_enrichseq(enrichseq_dir: Path, delim=\",\", output_file='taxid_abundances.csv'):\n",
    "    \"\"\"\n",
    "    Description\n",
    "    -----------\n",
    "        This function parses the enrichseq output directory.\n",
    "        \n",
    "    Note\n",
    "    ----------\n",
    "    use 'taxid_abundances.csv' or 'merge_overlap_out_refined.csv'\n",
    "    Input\n",
    "    -----------\n",
    "        1. Path to a singular EnrichSeq output directory.\n",
    "        2. [optional; Def=','] - delimiter\n",
    "        \n",
    "    Output\n",
    "    -----------\n",
    "        Abundance Dictionary containing:\n",
    "            1. taxid abundances\n",
    "            abundances_dict = {'tax_id_1' : 33,\n",
    "                               'tax_id_2' : 33,\n",
    "                               ...,\n",
    "                               'tax_id_N' : 33,\n",
    "                              }\n",
    "            2. cluster abundances\n",
    "            abundances_dict = {'cluster_1' : 33,\n",
    "                               'cluster_2' : 33,\n",
    "                               ...,\n",
    "                               'cluster_M' : 33,\n",
    "                              }\n",
    "    \"\"\"\n",
    "    # nested dictionary of different abundance measurements\n",
    "    abundances = {\"taxid_abundance\" : {},\"cluster_abundance\" : {}} \n",
    "    # get paths\n",
    "    output_path1 = enrichseq_dir / Path('enrichseq/output_files/')\n",
    "    output_path = enrichseq_dir / Path('enrichseq/output_files/')\n",
    "    taxid_abundances_path = output_path / Path(output_file)\n",
    "    \n",
    "    cluster_abundances_path = output_path1 / Path('cluster_abundances.csv')\n",
    "    csvs_to_parse = {\"taxid_abundance\" : taxid_abundances_path, \n",
    "                     \"cluster_abundance\" : cluster_abundances_path}\n",
    "    # parse abundance CSVs\n",
    "    for csv_name, csv_file in csvs_to_parse.items():\n",
    "        with open(csv_file, \"r\") as csv_file_opened:\n",
    "            csv_file_lines = csv_file_opened.readlines() # never that big, read all into RAM\n",
    "            for line in csv_file_lines:\n",
    "                tax_clust_id, abundance_val = line.strip(\"\\n\").split(delim)\n",
    "                abundances[csv_name][tax_clust_id] = float(abundance_val)\n",
    "\n",
    "    return abundances\n",
    "\n",
    "def parser_fastviromeexplorer(directory, name_column=0, counts_column=3, delim=\"\\t\"):\n",
    "    \"\"\"                          \n",
    "    Description\n",
    "    -----------\n",
    "        This function parses the FastViromeExplorer output directory.\n",
    "        \n",
    "    Input\n",
    "    -----------\n",
    "        1. Path to a singular FastViromeExplorer output directory.\n",
    "        2. column index corresponding to NCBI id\n",
    "        3. column index corresponding to the read counts\n",
    "        4. [optional; Def='\\t'] - delimiter\n",
    "        \n",
    "    Output\n",
    "    -----------\n",
    "        Abundance Dictionary containing:\n",
    "            1. taxid abundances\n",
    "            abundances_dict = {'tax_id_1' : 33,\n",
    "                               'tax_id_2' : 33,\n",
    "                               ...,\n",
    "                               'tax_id_N' : 33,\n",
    "                              }\n",
    "    \"\"\"\n",
    "    # init datastructures\n",
    "    abundances = {\"taxid_abundance\" : {}} # <tax/clust>id : abundance\n",
    "\n",
    "    # read in the NCBI to Tax id namings\n",
    "    with open('out_dictionary.txt') as f:\n",
    "        dictionary_str = f.read()\n",
    "    dictionary = json.loads(dictionary_str)\n",
    "    dictionary.update(make_dir())\n",
    "    dictionary.update(create_mapping_dictionary())\n",
    "    # get paths\n",
    "    taxid_abundances = directory / Path('abundance.tsv')\n",
    "    # parse abundance CSVs\n",
    "    csvs_to_parse = {\"taxid_abundance\" : taxid_abundances}\n",
    "    total_abundance = 0\n",
    "    for csv_name, csv_file in csvs_to_parse.items():\n",
    "        with open(csv_file, \"r\") as csv_file_opened:\n",
    "            csv_file_lines = csv_file_opened.readlines() # never that big, read all into RAM\n",
    "            for line_ind, line in enumerate(csv_file_lines):\n",
    "                if line_ind > 0: # skip header\n",
    "                    ncbi_id = line.strip(\"\\n\").split(delim)[name_column]\n",
    "                    counts_val = line.strip(\"\\n\").split(delim)[counts_column]\n",
    "\n",
    "                    # convert NCBI name to taxid, save counts\n",
    "                    if ncbi_id in dictionary:\n",
    "                        tax_clust_id = dictionary[ncbi_id] # 'dictionary' imported above\n",
    "                    elif ncbi_id.split(\".\")[0] in dictionary:\n",
    "                        tax_clust_id = dictionary[ncbi_id.split(\".\")[0]] # 'dictionary' imported above\n",
    "                    else:\n",
    "                        tax_clust_id = 'unk'\n",
    "                        if float(counts_val) > 0.01:\n",
    "                            print(f\"ERROR:\")\n",
    "                            print(f\"\\tFVE: {ncbi_id} NOT CONVERTED\")\n",
    "                            print(f\"\\tabundances: {float(counts_val)}\")\n",
    "                    abundances[csv_name][tax_clust_id] = float(counts_val)\n",
    "                    total_abundance += float(counts_val)\n",
    "    \n",
    "    # update abundances dictionary\n",
    "    ids_to_del = []\n",
    "    for tax_id, abundance_val in abundances[\"taxid_abundance\"].items():\n",
    "        if total_abundance > 0:\n",
    "            abundances[\"taxid_abundance\"][tax_id] /= total_abundance #normalize\n",
    "        if abundances[\"taxid_abundance\"][tax_id] < 0.001:\n",
    "            ids_to_del.append(tax_id)\n",
    "    # delete 0 value taxids\n",
    "    for val in ids_to_del:\n",
    "        del abundances[\"taxid_abundance\"][val]\n",
    "\n",
    "    return abundances\n",
    "\n",
    "def parser_bracken(directory, taxid_column=1, abundance_column=6, delim=\"\\t\"):\n",
    "    \"\"\"\n",
    "    Description\n",
    "    -----------\n",
    "        This function parses the enrichseq output directory.\n",
    "        \n",
    "    Input\n",
    "    -----------\n",
    "        1. Path to a singular Bracken output directory.\n",
    "        2. [optional; Def='\\t'] - delimiter\n",
    "        \n",
    "    Output\n",
    "    -----------\n",
    "        Abundance Dictionary containing:\n",
    "            1. taxid abundances\n",
    "            abundances_dict = {'tax_id_1' : 33,\n",
    "                               'tax_id_2' : 33,\n",
    "                               ...,\n",
    "                               'tax_id_N' : 33,\n",
    "                              }\n",
    "    \"\"\"\n",
    "    # dictionary of bracken abundance \n",
    "    abundances = {'taxid_abundance' : {}}\n",
    "    # get paths\n",
    "    taxid_abundances = directory / Path('abundances.bracken')\n",
    "    dictionary = make_dir()\n",
    "    files_to_parse = {\"taxid_abundance\" : taxid_abundances}\n",
    "    \n",
    "    # parse bracken abundance files\n",
    "    for file_name, bracken_file in files_to_parse.items():\n",
    "        with open(bracken_file, \"r\") as bracken_file_opened:\n",
    "            bracken_file_lines = bracken_file_opened.readlines() # never that big, read all into RAM\n",
    "            for line_ind, line in enumerate(bracken_file_lines):\n",
    "                if line_ind > 0: # skip header\n",
    "#                     ncbi_id = line.strip(\"\\n\").split(delim)[name_column]\n",
    "#                     counts_val = line.strip(\"\\n\").split(delim)[counts_column]\n",
    "                    tax_id = line.strip(\"\\n\").split(delim)[taxid_column]\n",
    "                    abundance_val = line.strip(\"\\n\").split(delim)[abundance_column]\n",
    "                    abundances[file_name][tax_id] = float(abundance_val)\n",
    "    \n",
    "    # update abundances dictionary\n",
    "    ids_to_del = []\n",
    "    for tax_id, abundance_val in abundances[\"taxid_abundance\"].items():\n",
    "        if abundances[\"taxid_abundance\"][tax_id] < 0.001:\n",
    "            ids_to_del.append(tax_id)\n",
    "    # delete 0 value taxids\n",
    "    for val in ids_to_del:\n",
    "        del abundances[\"taxid_abundance\"][val]\n",
    "\n",
    "    return abundances\n",
    "\n",
    "def make_dir():\n",
    "    ncbi2taxid = \"//scratch/summit/dral3008/BENCHMARKING_RESULTS/database/krakenDB/seqid2taxid.map\"\n",
    "    dictionary = {}\n",
    "    with open(ncbi2taxid) as ncbi2taxid:\n",
    "        lines = ncbi2taxid.readlines()\n",
    "        for line in lines:\n",
    "            ncbi, taxid = line.replace(\"\\n\",\"\").split(\"\\t\")\n",
    "            dictionary[ncbi] = taxid\n",
    "    return dictionary\n",
    "\n",
    "def parse_simulated_fasta(fasta_path):\n",
    "    \"\"\"                          \n",
    "    Description\n",
    "    -----------\n",
    "        This function parses the simulated true fasta file used\n",
    "        for testing.\n",
    "        \n",
    "    Input\n",
    "    -----------\n",
    "        1. Path to a singular simulated fasta file\n",
    "        \n",
    "    Output\n",
    "    -----------\n",
    "        Abundance Dictionary containing:\n",
    "            1. taxid abundances\n",
    "            abundances_dict = {'tax_id_1' : 33,\n",
    "                               'tax_id_2' : 33,\n",
    "                               ...,\n",
    "                               'tax_id_N' : 33,\n",
    "                              }\n",
    "    \"\"\"\n",
    "    abundances = {'taxid_abundance' : {}}\n",
    "    total_counts = 0\n",
    "    dictionary2 = make_dir()\n",
    "    not_found = set()\n",
    "    prev_len = 0\n",
    "    with open(fasta_path, \"r\") as fasta_opened:\n",
    "        line = fasta_opened.readline()\n",
    "        while (line):\n",
    "            if (line[0] == \">\"):\n",
    "                ncbi_id = line[1:].split(\"|\")[0].split(\"-\")[0]\n",
    "                # dictioinary from FastViromeExplorer, obtained above\n",
    "                if ncbi_id in dictionary2:\n",
    "                    name = dictionary2[line[1:].split(\"|\")[0].split(\"-\")[0]]\n",
    "                else:\n",
    "                    file_name = line[1:].split(\"|\")[-1].strip(\"\\n\")\n",
    "                    path = Path(\"//scratch/summit/dral3008/BENCHMARKING_RESULTS/database/ref_genomes/\" + file_name)\n",
    "                    try:\n",
    "                        name = open(path).readline().split(\"kraken:taxid|\")[1].strip(\"\\n\").replace(\" \", \"\")\n",
    "                    except:\n",
    "                        not_found.add(path)\n",
    "                        new_len = len(not_found)\n",
    "                        if new_len != prev_len:\n",
    "                            print(f\"number not found: {new_len}\")\n",
    "                            prev_len = new_len\n",
    "                # add name to dictionary\n",
    "#                 if name not in dictionary:\n",
    "#                     dictionary[ncbi_id] = name\n",
    "#                     dictionary[ncbi_id.split(\".\")[0]] = name\n",
    "# #                     print(ncbi_id)\n",
    "                # increment abundance level per read count\n",
    "                if name in abundances['taxid_abundance']:\n",
    "                    abundances['taxid_abundance'][name] += 1\n",
    "                else:\n",
    "                    abundances['taxid_abundance'][name] = 1\n",
    "                total_counts += 1\n",
    "            line = fasta_opened.readline()\n",
    "            \n",
    "    #turn counts into abundances - normalize\n",
    "    for tax_id in abundances['taxid_abundance'].keys():\n",
    "        abundances['taxid_abundance'][tax_id] /= total_counts #normalize\n",
    "    return abundances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # testing parsing functions\n",
    "\n",
    "# # Paths\n",
    "# fastvirome_test_dir = \"//scratch/summit/dral3008/benchmarking_enrichseq/results/FastViromeExplorer/num_mutations_test/3_genomes_sim_05del_05ins_illumina/\"\n",
    "# enrichseq_test_dir = \"//scratch/summit/dral3008/benchmarking_enrichseq/results/enrichseq/num_mutations_test/3_genomes_sim_00del_05ins_illumina./\"\n",
    "# truth_test_dir = \"//scratch/summit/dral3008/benchmarking_enrichseq/tests/num_mutations_test/3_genomes_sim_05del_05ins_illumina.fa\"\n",
    "# # run tests\n",
    "# fastvirome_values = parser_fastviromeexplorer(fastvirome_test_dir)\n",
    "# enrichseq_values = parser_enrichseq(enrichseq_test_dir)\n",
    "# true_values = parse_simulated_fasta(truth_test_dir)\n",
    "# print(fastvirome_values)\n",
    "# print(enrichseq_values)\n",
    "# print(true_values)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Results Methods and structures\n",
    "The methods and data structures here are used for creating a common data structure for the output of all tools being compared with one another. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "class ResultStruct:\n",
    "    \"\"\" EnrichSeq - datastruct for holding results \"\"\"\n",
    "    \n",
    "    def __init__(self):\n",
    "        self.l2_abundance_distance = []\n",
    "        self.classification_recall = 0\n",
    "        self.classification_precision = 0    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "PARSERS_MAP = {\"EnrichSeq_k75m31\" : parser_enrichseq,\n",
    "               \"EnrichSeq_k28m28\" : parser_enrichseq,\n",
    "               \"EnrichSeq_k35m31\" : parser_enrichseq,\n",
    "               \"EnrichSeq_k25\" : parser_enrichseq,\n",
    "               \"EnrichSeq\" : parser_enrichseq,\n",
    "               \"EnrichSeq_wMergeOverlap\" : parser_enrichseq,\n",
    "               \"FastViromeExplorer\" : parser_fastviromeexplorer,\n",
    "               \"Bracken\" : parser_bracken,\n",
    "               \"truth\" : parse_simulated_fasta}\n",
    "\n",
    "def parse_dir(results_directory, test_directory, parsers=PARSERS_MAP):\n",
    "    \"\"\"                          \n",
    "    Description\n",
    "    -----------\n",
    "        This function parses all of the result directories including\n",
    "        the test directory and returns a data structure containing the results.\n",
    "        \n",
    "    Input\n",
    "    -----------\n",
    "        1. Path to a singular simulated fasta file\n",
    "        \n",
    "    Output\n",
    "    -----------\n",
    "        Abundance Dictionary containing:\n",
    "            1. taxid abundances\n",
    "            abundances_dict = {'tax_id_1' : 33,\n",
    "                               'tax_id_2' : 33,\n",
    "                               ...,\n",
    "                               'tax_id_N' : 33,\n",
    "                              }\n",
    "    \"\"\"\n",
    "    results = {} # final structure\n",
    "    \n",
    "    # step 1: parse truth\n",
    "    results = parse_true_directory(test_directory, results, parsers)\n",
    "    \n",
    "    # step 2: parse results\n",
    "    for tool_name in os.listdir(results_directory):\n",
    "        results_path = Path(results_directory) / Path(tool_name)\n",
    "        results = parse_true_directory(results_path, results, parsers, tool_name=tool_name)\n",
    "    \n",
    "    # step 3: save raw data in csv\n",
    "#     print(results)\n",
    "    \n",
    "    # create results from parsed information\n",
    "    results = parsed_dict_to_results(results)\n",
    "    \n",
    "    return results\n",
    "    \n",
    "def parse_true_directory(test_directory, results: Dict, parsers, tool_name='true'):\n",
    "    \"\"\"                          \n",
    "    Description\n",
    "    -----------\n",
    "        This function parses the truth directory, adding the parsed \n",
    "        abundance results to input results dictionary.\n",
    "        \n",
    "                    TEST\n",
    "                      |\n",
    "                    /   \\\n",
    "                TEST_1    TEST_2\n",
    "               /   |        |    \\\n",
    "       test-1.a test-1.b test-2.a test-2.b\n",
    "       \n",
    "                    into \n",
    "                       \n",
    "                   Results (a dictionary)\n",
    "                      |\n",
    "                    /   \\\n",
    "             \"TEST_1\"    \"TEST_2\"\n",
    "               /   |        |    \\\n",
    "    trueAb-1.a trueAb-1.b trueAb-2.a trueAb-2.b\n",
    "         \n",
    "    Input\n",
    "    -----------\n",
    "        1. Path to a the entire truth test directory\n",
    "        2. results dictionary\n",
    "        3. parsers\n",
    "        \n",
    "    Output\n",
    "    -----------\n",
    "        results dictionary with the true abundance values added:\n",
    "        \n",
    "        example:\n",
    "        'num_genomes_test': {\n",
    "                             '5_genomes_sim_illumina': {'true': \n",
    "                                  {'taxid_abundance': {'2886930': 0.2, \n",
    "                                                       '10868': 0.2, \n",
    "                                                       '2681618': 0.2, \n",
    "                                                       '10658': 0.2, \n",
    "                                                       '127507': 0.2}\n",
    "                                  }\n",
    "                             }\n",
    "\n",
    "    \"\"\"\n",
    "    # parse truth\n",
    "    for test_directory_name in os.listdir(test_directory):\n",
    "        test_full_path = Path(test_directory) / Path(test_directory_name) \n",
    "        if os.path.isdir(test_full_path): # if directory [should be directory]\n",
    "            # create sub dictionary if not exists\n",
    "            if test_directory_name not in results: results[test_directory_name] = {}\n",
    "            # loop through all fasta files and parse for true abundance\n",
    "            for file_test in os.listdir(test_full_path):\n",
    "                file_full_path = Path(test_full_path) / Path(file_test)\n",
    "                if os.path.isfile(file_full_path) and (file_full_path.suffix == \".fa\"): # if file and fasta\n",
    "                    # create subsub dictionary if not exists\n",
    "                    if file_test not in results[test_directory_name]:\n",
    "                        file_test = file_test.replace(\".fa\", \"\") # get name from file\n",
    "                        results[test_directory_name][file_test] = {}\n",
    "                        \n",
    "                    # add abundances to results\n",
    "                    abundances = parse_simulated_fasta(file_full_path)\n",
    "                    # update the results dictionary\n",
    "                    results[test_directory_name][file_test][\"truth\"] = abundances\n",
    "                \n",
    "                if os.path.isdir(file_full_path): # is results dir\n",
    "                    # add abundances to results\n",
    "                    try:\n",
    "                        abundances = parsers[tool_name](file_full_path) \n",
    "                        results[test_directory_name][file_test.strip(\".\")][tool_name] = abundances\n",
    "                    except:\n",
    "                        print(f\"{tool_name} is missing {file_test} from {test_directory_name}\")\n",
    "                        continue\n",
    "    return results\n",
    "\n",
    "def parsed_dict_to_results(results):\n",
    "    \"\"\"                          \n",
    "    Description\n",
    "    -----------\n",
    "        This function parses the parsed results dictionary,\n",
    "        and creates a dictionary with all of the correct results.\n",
    "        Overall this utilizes the results structure.\n",
    "         \n",
    "    Input\n",
    "    -----------\n",
    "        1. Path to a results dictionary containing truth and \n",
    "           tool predictions\n",
    "        \n",
    "    Output\n",
    "    -----------\n",
    "        1. results dictionary with the tool results.\n",
    "    \"\"\"\n",
    "    for test_name in results.keys():\n",
    "        for sub_test in results[test_name].keys():\n",
    "            for tool_name, result_val in results[test_name][sub_test].items():\n",
    "                if tool_name != \"truth\":\n",
    "                    results[test_name][sub_test][tool_name] = get_results(results[test_name][sub_test][tool_name], \n",
    "                                                                          results[test_name][sub_test][\"truth\"])\n",
    "            del results[test_name][sub_test][\"truth\"]\n",
    "    return results\n",
    "\n",
    "def get_results(predicted_results, true_results):\n",
    "    \"\"\"                          \n",
    "    Description\n",
    "    -----------\n",
    "        Given the true results and tool results, this\n",
    "        function returns a Result Struct for a given tool.\n",
    "        \n",
    "    **NOTE**: This is where all of the metrics are calculated\n",
    "              so this is the most important function for the \n",
    "              benchmarking - contains the logic.\n",
    "         \n",
    "    Input\n",
    "    -----------\n",
    "        1. dictionary of predicted results\n",
    "        2. dictionary of true results\n",
    "        \n",
    "    Output\n",
    "    -----------\n",
    "        1. ResultStruct() with the tool results.\n",
    "    \"\"\"\n",
    "    # init\n",
    "    results_structure = ResultStruct()\n",
    "    predicted_abundances = predicted_results['taxid_abundance']\n",
    "    true_abundances = true_results['taxid_abundance']\n",
    "    \n",
    "    # error handling\n",
    "    if (len(true_abundances) == 0): \n",
    "        print(f\"TRUTH EMPTY\")\n",
    "        return results_structure\n",
    "    \n",
    "    # classification results\n",
    "    ## true positives\n",
    "    true_positives = []\n",
    "    for pred_taxid in predicted_abundances.keys():\n",
    "        if pred_taxid in true_abundances.keys():\n",
    "            true_positives.append(pred_taxid)\n",
    "    ## false positives\n",
    "    false_positives = []\n",
    "    for pred_taxid in predicted_abundances.keys():\n",
    "        if (pred_taxid not in true_abundances.keys()) and (pred_taxid != \"UNK\"):\n",
    "            false_positives.append(pred_taxid)\n",
    "    ## false negatives\n",
    "    false_negatives = []\n",
    "    for pred_taxid in true_abundances.keys():\n",
    "        if pred_taxid not in predicted_abundances.keys():\n",
    "            false_negatives.append(pred_taxid)\n",
    "            \n",
    "    ## get metrics\n",
    "    recall = len(true_positives) / (len(true_positives) + len(false_negatives))\n",
    "    precision = len(true_positives) / (len(true_positives) + len(false_positives))\n",
    "    \n",
    "    # abundance results\n",
    "    ## L2 distances\n",
    "    l2_dist_array = []\n",
    "    total_abundance_wo_unk = sum([abundance for key, abundance in predicted_abundances.items() if key != \"UNK\" ])\n",
    "    for tax_id, pred_abundance in predicted_abundances.items():\n",
    "        if tax_id in true_abundances:\n",
    "            pred_abundance /= total_abundance_wo_unk\n",
    "            l2_diff = abs(pred_abundance - true_abundances[tax_id])**2\n",
    "        else:\n",
    "            l2_diff = 0\n",
    "        l2_dist_array.append(l2_diff)\n",
    "    \n",
    "    # store into results structure\n",
    "    results_structure.l2_abundance_distance = l2_dist_array\n",
    "    results_structure.classification_recall = recall\n",
    "    results_structure.classification_precision = precision\n",
    "    \n",
    "    return results_structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number not found: 1\n",
      "number not found: 1\n",
      "number not found: 1\n",
      "ERROR:\n",
      "\tFVE: NC_029069.1 NOT CONVERTED\n",
      "\tabundances: 1.16102\n"
     ]
    }
   ],
   "source": [
    "# create 'results'\n",
    "results = parse_dir(\"//scratch/summit/dral3008/benchmarking_enrichseq/RESULTS/results-NEW/test_1/\",\n",
    "                    \"//scratch/summit/dral3008/benchmarking_enrichseq/TESTS/tests_new/test_1/\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TO CSV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "replicate_ID,MergeOverlap,major_test,sub_test,tool,genome_count,read_count,classification_recall,classification_precision,l2_abundance_distance\n",
      "number not found: 1\n",
      "number not found: 1\n",
      "number not found: 1\n",
      "ERROR:\n",
      "\tFVE: NC_029069.1 NOT CONVERTED\n",
      "\tabundances: 1.16102\n",
      "0,False,num_genomes,800_genomes_500000_reads,FastViromeExplorer,800,500000,0.48,0.8322147651006712,1.1559058055522562e-06\n",
      "0,False,num_genomes,800_genomes_500000_reads,EnrichSeq_k75m31,800,500000,0.7148387096774194,0.8835725677830941,1.5171363494827102e-06\n",
      "0,False,num_genomes,800_genomes_500000_reads,EnrichSeq_k35m31,800,500000,0.7354838709677419,0.8444444444444444,1.3676581907537333e-06\n",
      "0,False,num_genomes,800_genomes_500000_reads,EnrichSeq_k28m28,800,500000,0.7303225806451613,0.5901981230448383,5.438687908528633e-07\n",
      "0,False,num_genomes,800_genomes_500000_reads,Bracken,800,500000,0.29419354838709677,0.5428571428571428,8.438969580623027e-07\n",
      "0,False,num_genomes,200_genomes_500000_reads,FastViromeExplorer,200,500000,0.52,0.6153846153846154,5.376971332510971e-06\n",
      "0,False,num_genomes,200_genomes_500000_reads,EnrichSeq_k75m31,200,500000,0.925,0.8564814814814815,3.520511771208775e-06\n",
      "0,False,num_genomes,200_genomes_500000_reads,EnrichSeq_k35m31,200,500000,0.925,0.8149779735682819,3.439088917733069e-06\n",
      "0,False,num_genomes,200_genomes_500000_reads,EnrichSeq_k28m28,200,500000,0.925,0.4946524064171123,1.8470420827095384e-06\n",
      "0,False,num_genomes,200_genomes_500000_reads,Bracken,200,500000,0.31,0.38509316770186336,4.267764107275761e-06\n",
      "0,False,num_genomes,400_genomes_500000_reads,FastViromeExplorer,400,500000,0.4625,0.6751824817518248,2.2593615651573903e-06\n",
      "0,False,num_genomes,400_genomes_500000_reads,EnrichSeq_k75m31,400,500000,0.7925,0.9109195402298851,4.3588965423018515e-06\n",
      "0,False,num_genomes,400_genomes_500000_reads,EnrichSeq_k35m31,400,500000,0.815,0.8556430446194225,3.8115797369083757e-06\n",
      "0,False,num_genomes,400_genomes_500000_reads,EnrichSeq_k28m28,400,500000,0.805,0.5865209471766849,1.897809084193383e-06\n",
      "0,False,num_genomes,400_genomes_500000_reads,Bracken,400,500000,0.2725,0.4144486692015209,1.9102018105382983e-06\n",
      "0,False,num_genomes,1000_genomes_500000_reads,FastViromeExplorer,1000,500000,0.4774193548387097,0.8185840707964602,9.258926891191184e-07\n",
      "0,False,num_genomes,1000_genomes_500000_reads,EnrichSeq_k75m31,1000,500000,0.4890322580645161,0.8981042654028436,5.916941751479552e-06\n",
      "0,False,num_genomes,1000_genomes_500000_reads,EnrichSeq_k35m31,1000,500000,0.5329032258064517,0.8731501057082452,4.288106312752809e-06\n",
      "0,False,num_genomes,1000_genomes_500000_reads,EnrichSeq_k28m28,1000,500000,0.512258064516129,0.6672268907563025,1.7136463469077262e-06\n",
      "0,False,num_genomes,1000_genomes_500000_reads,Bracken,1000,500000,0.29161290322580646,0.5280373831775701,9.209463782941376e-07\n",
      "0,False,num_genomes,600_genomes_500000_reads,FastViromeExplorer,600,500000,0.4607679465776294,0.7340425531914894,1.111750121887458e-06\n",
      "0,False,num_genomes,600_genomes_500000_reads,EnrichSeq_k75m31,600,500000,0.6343906510851419,0.9026128266033254,4.568756081991019e-06\n",
      "0,False,num_genomes,600_genomes_500000_reads,EnrichSeq_k35m31,600,500000,0.672787979966611,0.8466386554621849,3.42219475897289e-06\n",
      "0,False,num_genomes,600_genomes_500000_reads,EnrichSeq_k28m28,600,500000,0.654424040066778,0.6182965299684543,1.0679638506212607e-06\n",
      "0,False,num_genomes,600_genomes_500000_reads,Bracken,600,500000,0.2888146911519199,0.4818941504178273,9.745302606359793e-07\n",
      "0,False,num_genomes,10_genomes_500000_reads,FastViromeExplorer,10,500000,0.4,0.0851063829787234,0.00046197900207304996\n",
      "0,False,num_genomes,10_genomes_500000_reads,EnrichSeq_k75m31,10,500000,0.9,1.0,0.0001111119299828682\n",
      "0,False,num_genomes,10_genomes_500000_reads,EnrichSeq_k35m31,10,500000,0.9,1.0,0.0001111119299828682\n",
      "0,False,num_genomes,10_genomes_500000_reads,EnrichSeq_k28m28,10,500000,0.7,0.6363636363636364,0.00023324768469264534\n",
      "0,False,num_genomes,10_genomes_500000_reads,Bracken,10,500000,0.2,0.041666666666666664,0.00023364463259201921\n",
      "0,False,num_reads_800genomes,800_genomes_500000_reads,FastViromeExplorer,800,500000,0.52,0.6190476190476191,5.474658781045473e-06\n",
      "0,False,num_reads_800genomes,800_genomes_500000_reads,EnrichSeq_k75m31,800,500000,0.915,0.8672985781990521,3.93120882734197e-06\n",
      "0,False,num_reads_800genomes,800_genomes_500000_reads,EnrichSeq_k35m31,800,500000,0.92,0.8401826484018264,3.6498696472421207e-06\n",
      "0,False,num_reads_800genomes,800_genomes_500000_reads,EnrichSeq_k28m28,800,500000,0.9,0.5278592375366569,1.8697374007940021e-06\n",
      "0,False,num_reads_800genomes,800_genomes_500000_reads,Bracken,800,500000,0.31,0.37575757575757573,3.965788980866646e-06\n",
      "0,False,num_reads_800genomes,800_genomes_400000_reads,FastViromeExplorer,800,400000,0.52,0.6046511627906976,5.219811032347403e-06\n",
      "0,False,num_reads_800genomes,800_genomes_400000_reads,EnrichSeq_k75m31,800,400000,0.885,0.8805970149253731,4.353180023472515e-06\n",
      "0,False,num_reads_800genomes,800_genomes_400000_reads,EnrichSeq_k35m31,800,400000,0.895,0.7955555555555556,3.724557483165449e-06\n",
      "0,False,num_reads_800genomes,800_genomes_400000_reads,EnrichSeq_k28m28,800,400000,0.885,0.5566037735849056,2.05042100209794e-06\n",
      "0,False,num_reads_800genomes,800_genomes_400000_reads,Bracken,800,400000,0.31,0.38271604938271603,4.163891382014149e-06\n",
      "0,False,num_reads_800genomes,800_genomes_300000_reads,FastViromeExplorer,800,300000,0.52,0.6153846153846154,5.4421046363511174e-06\n",
      "0,False,num_reads_800genomes,800_genomes_300000_reads,EnrichSeq_k75m31,800,300000,0.83,0.8736842105263158,5.906776084298838e-06\n",
      "0,False,num_reads_800genomes,800_genomes_300000_reads,EnrichSeq_k35m31,800,300000,0.845,0.8492462311557789,5.3733110889244385e-06\n",
      "0,False,num_reads_800genomes,800_genomes_300000_reads,EnrichSeq_k28m28,800,300000,0.825,0.5892857142857143,2.500536746707111e-06\n",
      "0,False,num_reads_800genomes,800_genomes_300000_reads,Bracken,800,300000,0.305,0.3836477987421384,5.155161498727848e-06\n",
      "0,False,num_reads_800genomes,800_genomes_100000_reads,FastViromeExplorer,800,100000,0.52,0.6081871345029239,5.20510699030386e-06\n",
      "0,False,num_reads_800genomes,800_genomes_100000_reads,EnrichSeq_k75m31,800,100000,0.355,0.9342105263157895,9.643923175253288e-05\n",
      "0,False,num_reads_800genomes,800_genomes_100000_reads,EnrichSeq_k35m31,800,100000,0.385,0.9390243902439024,7.632128365068736e-05\n",
      "0,False,num_reads_800genomes,800_genomes_100000_reads,EnrichSeq_k28m28,800,100000,0.375,0.7894736842105263,5.082740512546891e-05\n",
      "0,False,num_reads_800genomes,800_genomes_100000_reads,Bracken,800,100000,0.295,0.3597560975609756,5.786456368045979e-06\n",
      "0,False,num_reads_800genomes,800_genomes_900000_reads,FastViromeExplorer,800,900000,0.52,0.6227544910179641,5.478328692880774e-06\n",
      "0,False,num_reads_800genomes,800_genomes_900000_reads,EnrichSeq_k75m31,800,900000,0.93,0.8493150684931506,3.5080001370579448e-06\n",
      "0,False,num_reads_800genomes,800_genomes_900000_reads,EnrichSeq_k35m31,800,900000,0.935,0.8237885462555066,3.4098092116940224e-06\n",
      "0,False,num_reads_800genomes,800_genomes_900000_reads,EnrichSeq_k28m28,800,900000,0.92,0.4791666666666667,1.889116498163153e-06\n",
      "0,False,num_reads_800genomes,800_genomes_900000_reads,Bracken,800,900000,0.31,0.3803680981595092,4.104277423239665e-06\n",
      "0,False,num_reads_800genomes,800_genomes_600000_reads,FastViromeExplorer,800,600000,0.52,0.6153846153846154,5.366754499308622e-06\n",
      "0,False,num_reads_800genomes,800_genomes_600000_reads,EnrichSeq_k75m31,800,600000,0.92,0.8598130841121495,3.72438720335674e-06\n",
      "0,False,num_reads_800genomes,800_genomes_600000_reads,EnrichSeq_k35m31,800,600000,0.925,0.8295964125560538,3.399537058040335e-06\n",
      "0,False,num_reads_800genomes,800_genomes_600000_reads,EnrichSeq_k28m28,800,600000,0.92,0.5027322404371585,1.8337138170713335e-06\n",
      "0,False,num_reads_800genomes,800_genomes_600000_reads,Bracken,800,600000,0.31,0.389937106918239,4.301587155776585e-06\n",
      "0,False,num_reads_800genomes,800_genomes_800000_reads,FastViromeExplorer,800,800000,0.52,0.6046511627906976,5.201963507305405e-06\n",
      "0,False,num_reads_800genomes,800_genomes_800000_reads,EnrichSeq_k75m31,800,800000,0.93,0.8571428571428571,3.4868392167392774e-06\n",
      "0,False,num_reads_800genomes,800_genomes_800000_reads,EnrichSeq_k35m31,800,800000,0.94,0.8663594470046083,3.5746316661634176e-06\n",
      "0,False,num_reads_800genomes,800_genomes_800000_reads,EnrichSeq_k28m28,800,800000,0.935,0.4675,1.7652711684358648e-06\n",
      "0,False,num_reads_800genomes,800_genomes_800000_reads,Bracken,800,800000,0.31,0.3875,4.164589864336873e-06\n",
      "0,False,num_reads_800genomes,800_genomes_700000_reads,FastViromeExplorer,800,700000,0.52,0.6190476190476191,5.4405838201807986e-06\n",
      "0,False,num_reads_800genomes,800_genomes_700000_reads,EnrichSeq_k75m31,800,700000,0.93,0.8193832599118943,3.467052805609227e-06\n",
      "0,False,num_reads_800genomes,800_genomes_700000_reads,EnrichSeq_k35m31,800,700000,0.93,0.8017241379310345,3.378755017991163e-06\n",
      "0,False,num_reads_800genomes,800_genomes_700000_reads,EnrichSeq_k28m28,800,700000,0.925,0.4894179894179894,1.9332304440169596e-06\n",
      "0,False,num_reads_800genomes,800_genomes_700000_reads,Bracken,800,700000,0.305,0.3765432098765432,5.034468332888709e-06\n",
      "0,False,num_reads_800genomes,800_genomes_1000000_reads,FastViromeExplorer,800,1000000,0.52,0.6190476190476191,5.427807305422274e-06\n",
      "0,False,num_reads_800genomes,800_genomes_1000000_reads,EnrichSeq_k75m31,800,1000000,0.92,0.8679245283018868,3.495159900607631e-06\n",
      "0,False,num_reads_800genomes,800_genomes_1000000_reads,EnrichSeq_k35m31,800,1000000,0.93,0.8340807174887892,3.3255228611092274e-06\n",
      "0,False,num_reads_800genomes,800_genomes_1000000_reads,EnrichSeq_k28m28,800,1000000,0.91,0.5041551246537396,1.8392893697936104e-06\n",
      "0,False,num_reads_800genomes,800_genomes_1000000_reads,Bracken,800,1000000,0.31,0.3875,4.049723472239673e-06\n",
      "0,False,num_reads_800genomes,800_genomes_200000_reads,FastViromeExplorer,800,200000,0.52,0.6081871345029239,5.313018736244159e-06\n",
      "0,False,num_reads_800genomes,800_genomes_200000_reads,EnrichSeq_k75m31,800,200000,0.77,0.9058823529411765,7.724436296697913e-06\n",
      "0,False,num_reads_800genomes,800_genomes_200000_reads,EnrichSeq_k35m31,800,200000,0.79,0.8404255319148937,6.421541174109512e-06\n",
      "0,False,num_reads_800genomes,800_genomes_200000_reads,EnrichSeq_k28m28,800,200000,0.755,0.5992063492063492,3.162847508194713e-06\n",
      "0,False,num_reads_800genomes,800_genomes_200000_reads,Bracken,800,200000,0.305,0.37888198757763975,4.037399124744544e-06\n",
      "0,False,num_reads_200genomes,200_genomes_500000_reads,FastViromeExplorer,200,500000,0.52,0.6046511627906976,5.245107844211835e-06\n",
      "0,False,num_reads_200genomes,200_genomes_500000_reads,EnrichSeq_k75m31,200,500000,0.92,0.8679245283018868,3.4443979619914746e-06\n",
      "0,False,num_reads_200genomes,200_genomes_500000_reads,EnrichSeq_k35m31,200,500000,0.925,0.8149779735682819,3.2148251137901416e-06\n",
      "0,False,num_reads_200genomes,200_genomes_500000_reads,EnrichSeq_k28m28,200,500000,0.92,0.5041095890410959,1.9230275197487102e-06\n",
      "0,False,num_reads_200genomes,200_genomes_500000_reads,Bracken,200,500000,0.31,0.3875,4.2428431600178674e-06\n",
      "0,False,num_reads_200genomes,200_genomes_100000_reads,FastViromeExplorer,200,100000,0.515,0.5919540229885057,5.05986294319273e-06\n",
      "0,False,num_reads_200genomes,200_genomes_100000_reads,EnrichSeq_k75m31,200,100000,0.385,0.9625,9.551774667077068e-05\n",
      "0,False,num_reads_200genomes,200_genomes_100000_reads,EnrichSeq_k35m31,200,100000,0.415,0.9120879120879121,6.865124154703334e-05\n",
      "0,False,num_reads_200genomes,200_genomes_100000_reads,EnrichSeq_k28m28,200,100000,0.335,0.6633663366336634,4.852127246099267e-05\n",
      "0,False,num_reads_200genomes,200_genomes_100000_reads,Bracken,200,100000,0.305,0.37423312883435583,5.440122851586563e-06\n",
      "0,False,num_reads_200genomes,200_genomes_800000_reads,FastViromeExplorer,200,800000,0.52,0.6153846153846154,5.452295901030558e-06\n",
      "0,False,num_reads_200genomes,200_genomes_800000_reads,EnrichSeq_k75m31,200,800000,0.93,0.8193832599118943,3.609087406709131e-06\n",
      "0,False,num_reads_200genomes,200_genomes_800000_reads,EnrichSeq_k35m31,200,800000,0.935,0.8025751072961373,3.4700982852711484e-06\n",
      "0,False,num_reads_200genomes,200_genomes_800000_reads,EnrichSeq_k28m28,200,800000,0.925,0.4755784061696658,1.8677228766816694e-06\n",
      "0,False,num_reads_200genomes,200_genomes_800000_reads,Bracken,200,800000,0.31,0.3803680981595092,3.918536785934714e-06\n",
      "0,False,num_reads_200genomes,200_genomes_300000_reads,FastViromeExplorer,200,300000,0.52,0.6153846153846154,5.454353010181274e-06\n",
      "0,False,num_reads_200genomes,200_genomes_300000_reads,EnrichSeq_k75m31,200,300000,0.81,0.8852459016393442,6.298668012970233e-06\n",
      "0,False,num_reads_200genomes,200_genomes_300000_reads,EnrichSeq_k28m28,200,300000,0.8,0.6153846153846154,3.145859358086071e-06\n",
      "0,False,num_reads_200genomes,200_genomes_300000_reads,Bracken,200,300000,0.3,0.36585365853658536,4.831818237286163e-06\n",
      "0,False,num_reads_200genomes,200_genomes_1000000_reads,FastViromeExplorer,200,1000000,0.52,0.6190476190476191,5.417608492914437e-06\n",
      "0,False,num_reads_200genomes,200_genomes_1000000_reads,EnrichSeq_k75m31,200,1000000,0.935,0.8779342723004695,3.4411071718349037e-06\n",
      "0,False,num_reads_200genomes,200_genomes_1000000_reads,EnrichSeq_k35m31,200,1000000,0.94,0.8034188034188035,3.1439115140472167e-06\n",
      "0,False,num_reads_200genomes,200_genomes_1000000_reads,EnrichSeq_k28m28,200,1000000,0.935,0.49340369393139843,1.9406257243102676e-06\n",
      "0,False,num_reads_200genomes,200_genomes_1000000_reads,Bracken,200,1000000,0.31,0.3803680981595092,3.9774466240961145e-06\n",
      "0,False,num_reads_200genomes,200_genomes_600000_reads,FastViromeExplorer,200,600000,0.52,0.6227544910179641,5.512144222588e-06\n",
      "0,False,num_reads_200genomes,200_genomes_600000_reads,EnrichSeq_k75m31,200,600000,0.915,0.8714285714285714,3.6433375877314e-06\n",
      "0,False,num_reads_200genomes,200_genomes_600000_reads,EnrichSeq_k35m31,200,600000,0.925,0.8295964125560538,3.316884684957524e-06\n",
      "0,False,num_reads_200genomes,200_genomes_600000_reads,EnrichSeq_k28m28,200,600000,0.915,0.4765625,1.7277402865918543e-06\n",
      "0,False,num_reads_200genomes,200_genomes_600000_reads,Bracken,200,600000,0.31,0.3803680981595092,4.318396164454907e-06\n",
      "0,False,num_reads_200genomes,200_genomes_200000_reads,FastViromeExplorer,200,200000,0.52,0.6190476190476191,5.487275384039745e-06\n",
      "0,False,num_reads_200genomes,200_genomes_200000_reads,EnrichSeq_k75m31,200,200000,0.765,0.9053254437869822,8.135134081442285e-06\n",
      "0,False,num_reads_200genomes,200_genomes_200000_reads,EnrichSeq_k35m31,200,200000,0.78,0.8764044943820225,7.431551256071917e-06\n",
      "0,False,num_reads_200genomes,200_genomes_200000_reads,EnrichSeq_k28m28,200,200000,0.765,0.6891891891891891,4.174779009486643e-06\n",
      "0,False,num_reads_200genomes,200_genomes_200000_reads,Bracken,200,200000,0.3,0.37267080745341613,4.442192438358081e-06\n",
      "0,False,num_reads_200genomes,200_genomes_700000_reads,FastViromeExplorer,200,700000,0.515,0.5988372093023255,5.209806373003963e-06\n",
      "0,False,num_reads_200genomes,200_genomes_700000_reads,EnrichSeq_k75m31,200,700000,0.93,0.8493150684931506,3.5818536041619876e-06\n",
      "0,False,num_reads_200genomes,200_genomes_700000_reads,EnrichSeq_k35m31,200,700000,0.935,0.8060344827586207,3.282954815707441e-06\n",
      "0,False,num_reads_200genomes,200_genomes_700000_reads,EnrichSeq_k28m28,200,700000,0.92,0.4730077120822622,1.9077447115982094e-06\n",
      "0,False,num_reads_200genomes,200_genomes_700000_reads,Bracken,200,700000,0.31,0.37349397590361444,3.861586272097531e-06\n",
      "0,False,num_reads_200genomes,200_genomes_400000_reads,FastViromeExplorer,200,400000,0.52,0.6153846153846154,5.3736039023714736e-06\n",
      "0,False,num_reads_200genomes,200_genomes_400000_reads,EnrichSeq_k75m31,200,400000,0.905,0.8829268292682927,3.908291022340644e-06\n",
      "0,False,num_reads_200genomes,200_genomes_400000_reads,EnrichSeq_k35m31,200,400000,0.915,0.8061674008810573,3.659988671705036e-06\n",
      "0,False,num_reads_200genomes,200_genomes_400000_reads,EnrichSeq_k28m28,200,400000,0.9,0.5309734513274337,1.8680225397865729e-06\n",
      "0,False,num_reads_200genomes,200_genomes_400000_reads,Bracken,200,400000,0.31,0.3803680981595092,3.89201051781403e-06\n",
      "0,False,num_reads_200genomes,200_genomes_900000_reads,FastViromeExplorer,200,900000,0.52,0.6153846153846154,5.452968508858343e-06\n",
      "0,False,num_reads_200genomes,200_genomes_900000_reads,EnrichSeq_k75m31,200,900000,0.935,0.8538812785388128,3.6093429011903567e-06\n",
      "0,False,num_reads_200genomes,200_genomes_900000_reads,EnrichSeq_k35m31,200,900000,0.93,0.8193832599118943,3.485467974956424e-06\n",
      "0,False,num_reads_200genomes,200_genomes_900000_reads,EnrichSeq_k28m28,200,900000,0.92,0.47668393782383417,1.7754643234030518e-06\n",
      "0,False,num_reads_200genomes,200_genomes_900000_reads,Bracken,200,900000,0.31,0.38271604938271603,4.194797319805089e-06\n",
      "number not found: 1\n",
      "number not found: 1\n",
      "FastViromeExplorer is missing 800_genomes_500000_reads from num_genomes\n",
      "EnrichSeq_k75m31 is missing 10_genomes_500000_reads from num_genomes\n",
      "EnrichSeq_k35m31 is missing 10_genomes_500000_reads from num_genomes\n",
      "EnrichSeq_k28m28 is missing 10_genomes_500000_reads from num_genomes\n",
      "EnrichSeq_k28m28 is missing 800_genomes_1000000_reads from num_reads_800genomes\n",
      "EnrichSeq_k28m28 is missing 800_genomes_800000_reads from num_reads_800genomes\n",
      "EnrichSeq_k28m28 is missing 800_genomes_900000_reads from num_reads_800genomes\n",
      "1,False,num_genomes,800_genomes_500000_reads,EnrichSeq_k75m31,800,500000,0.5097783572359843,0.9135514018691588,6.511622570856783e-06\n",
      "1,False,num_genomes,800_genomes_500000_reads,EnrichSeq_k35m31,800,500000,0.5488917861799217,0.864476386036961,5.238811482645343e-06\n",
      "1,False,num_genomes,800_genomes_500000_reads,Bracken,800,500000,0.29595827900912647,0.5550122249388753,9.288468801142137e-07\n",
      "1,False,num_genomes,200_genomes_500000_reads,FastViromeExplorer,200,500000,0.47,0.5081081081081081,4.236237560089545e-06\n",
      "1,False,num_genomes,200_genomes_500000_reads,EnrichSeq_k75m31,200,500000,0.91,0.8465116279069768,8.192883109000371e-06\n",
      "1,False,num_genomes,200_genomes_500000_reads,EnrichSeq_k35m31,200,500000,0.92,0.7763713080168776,7.43881990935777e-06\n",
      "1,False,num_genomes,200_genomes_500000_reads,Bracken,200,500000,0.31,0.34065934065934067,3.4739966641878764e-06\n",
      "1,False,num_genomes,400_genomes_500000_reads,FastViromeExplorer,400,500000,0.49874686716791977,0.7056737588652482,1.8725060682689142e-06\n",
      "1,False,num_genomes,400_genomes_500000_reads,EnrichSeq_k75m31,400,500000,0.7694235588972431,0.9136904761904762,3.191617161724325e-06\n",
      "1,False,num_genomes,400_genomes_500000_reads,EnrichSeq_k35m31,400,500000,0.7844611528822055,0.8575342465753425,2.809389996738097e-06\n",
      "1,False,num_genomes,400_genomes_500000_reads,Bracken,400,500000,0.3107769423558897,0.46096654275092935,1.5249617451570541e-06\n",
      "1,False,num_genomes,1000_genomes_500000_reads,FastViromeExplorer,1000,500000,0.46808510638297873,0.8494208494208494,8.201023903688899e-07\n",
      "1,False,num_genomes,1000_genomes_500000_reads,EnrichSeq_k75m31,1000,500000,0.42127659574468085,0.9230769230769231,6.867353388328956e-06\n",
      "1,False,num_genomes,1000_genomes_500000_reads,EnrichSeq_k35m31,1000,500000,0.46382978723404256,0.9121338912133892,5.472109169783717e-06\n",
      "1,False,num_genomes,1000_genomes_500000_reads,EnrichSeq_k28m28,1000,500000,0.44148936170212766,0.6814449917898193,2.9280451056057533e-06\n",
      "1,False,num_genomes,1000_genomes_500000_reads,Bracken,1000,500000,0.2925531914893617,0.5522088353413654,6.348948327098887e-07\n",
      "1,False,num_genomes,600_genomes_500000_reads,FastViromeExplorer,600,500000,0.445,0.7275204359673024,1.3312261342404896e-06\n",
      "1,False,num_genomes,600_genomes_500000_reads,EnrichSeq_k75m31,600,500000,0.6483333333333333,0.9174528301886793,4.735111530246436e-06\n",
      "1,False,num_genomes,600_genomes_500000_reads,EnrichSeq_k35m31,600,500000,0.68,0.8888888888888888,4.143604352700029e-06\n",
      "1,False,num_genomes,600_genomes_500000_reads,Bracken,600,500000,0.28833333333333333,0.4928774928774929,1.1820197285969298e-06\n",
      "1,False,num_reads_800genomes,800_genomes_500000_reads,EnrichSeq_k75m31,800,500000,0.5241199478487614,0.8874172185430463,6.5412091445053825e-06\n",
      "1,False,num_reads_800genomes,800_genomes_500000_reads,EnrichSeq_k35m31,800,500000,0.5658409387222947,0.8714859437751004,4.997557954616138e-06\n",
      "1,False,num_reads_800genomes,800_genomes_500000_reads,EnrichSeq_k28m28,800,500000,0.5267275097783573,0.6474358974358975,1.476208500388679e-06\n",
      "1,False,num_reads_800genomes,800_genomes_500000_reads,Bracken,800,500000,0.2920469361147327,0.5463414634146342,8.869944313295556e-07\n",
      "1,False,num_reads_800genomes,800_genomes_400000_reads,EnrichSeq_k75m31,800,400000,0.36766623207301175,0.9155844155844156,1.5105683958169278e-05\n",
      "1,False,num_reads_800genomes,800_genomes_400000_reads,EnrichSeq_k35m31,800,400000,0.4106910039113429,0.8923512747875354,1.1093812789642606e-05\n",
      "1,False,num_reads_800genomes,800_genomes_400000_reads,EnrichSeq_k28m28,800,400000,0.379400260756193,0.6326086956521739,6.663529394348254e-06\n",
      "1,False,num_reads_800genomes,800_genomes_400000_reads,Bracken,800,400000,0.2894393741851369,0.5495049504950495,9.955013904220536e-07\n",
      "1,False,num_reads_800genomes,800_genomes_300000_reads,EnrichSeq_k75m31,800,300000,0.32985658409387225,0.9133574007220217,1.761803174526725e-05\n",
      "1,False,num_reads_800genomes,800_genomes_300000_reads,EnrichSeq_k35m31,800,300000,0.37809647979139505,0.8978328173374613,1.4360872439291182e-05\n",
      "1,False,num_reads_800genomes,800_genomes_300000_reads,EnrichSeq_k28m28,800,300000,0.3376792698826597,0.6037296037296037,7.760520040481117e-06\n",
      "1,False,num_reads_800genomes,800_genomes_300000_reads,Bracken,800,300000,0.2894393741851369,0.5349397590361445,9.179456195737858e-07\n",
      "1,False,num_reads_800genomes,800_genomes_100000_reads,EnrichSeq_k75m31,800,100000,0.04563233376792699,0.875,0.0012442886708666683\n",
      "1,False,num_reads_800genomes,800_genomes_100000_reads,EnrichSeq_k35m31,800,100000,0.054758800521512385,0.7924528301886793,0.0007596313025648414\n",
      "1,False,num_reads_800genomes,800_genomes_100000_reads,EnrichSeq_k28m28,800,100000,0.0469361147327249,0.6428571428571429,0.0007621807546245963\n",
      "1,False,num_reads_800genomes,800_genomes_100000_reads,Bracken,800,100000,0.2777053455019557,0.5406091370558376,1.4627551579020575e-06\n",
      "1,False,num_reads_800genomes,800_genomes_900000_reads,EnrichSeq_k75m31,800,900000,0.758800521512386,0.8858447488584474,2.3266810066385155e-06\n",
      "1,False,num_reads_800genomes,800_genomes_900000_reads,EnrichSeq_k35m31,800,900000,0.7809647979139505,0.865606936416185,2.198698233866403e-06\n",
      "1,False,num_reads_800genomes,800_genomes_900000_reads,Bracken,800,900000,0.29465449804432853,0.5432692307692307,8.372612129554209e-07\n",
      "1,False,num_reads_800genomes,800_genomes_600000_reads,EnrichSeq_k75m31,800,600000,0.6232073011734028,0.9263565891472868,4.446096902208028e-06\n",
      "1,False,num_reads_800genomes,800_genomes_600000_reads,EnrichSeq_k35m31,800,600000,0.666232073011734,0.8705281090289608,3.615284357812605e-06\n",
      "1,False,num_reads_800genomes,800_genomes_600000_reads,EnrichSeq_k28m28,800,600000,0.6245110821382008,0.6378162450066578,2.474192030211417e-06\n",
      "1,False,num_reads_800genomes,800_genomes_600000_reads,Bracken,800,600000,0.29595827900912647,0.5483091787439613,9.669058922395975e-07\n",
      "1,False,num_reads_800genomes,800_genomes_800000_reads,EnrichSeq_k75m31,800,800000,0.7340286831812256,0.9095315024232633,2.598251388985994e-06\n",
      "1,False,num_reads_800genomes,800_genomes_800000_reads,EnrichSeq_k35m31,800,800000,0.7640156453715776,0.8668639053254438,2.3703429891168524e-06\n",
      "1,False,num_reads_800genomes,800_genomes_800000_reads,Bracken,800,800000,0.2985658409387223,0.5465393794749404,8.027190867540191e-07\n",
      "1,False,num_reads_800genomes,800_genomes_700000_reads,EnrichSeq_k75m31,800,700000,0.666232073011734,0.9060283687943262,3.176019354390851e-06\n",
      "1,False,num_reads_800genomes,800_genomes_700000_reads,EnrichSeq_k35m31,800,700000,0.7053455019556715,0.8811074918566775,2.9408169034506705e-06\n",
      "1,False,num_reads_800genomes,800_genomes_700000_reads,EnrichSeq_k28m28,800,700000,0.6844850065189049,0.6410256410256411,1.9216870882940152e-06\n",
      "1,False,num_reads_800genomes,800_genomes_700000_reads,Bracken,800,700000,0.29465449804432853,0.5432692307692307,8.539993378706401e-07\n",
      "1,False,num_reads_800genomes,800_genomes_1000000_reads,EnrichSeq_k75m31,800,1000000,0.7770534550195567,0.8855869242199108,2.2428828561272343e-06\n",
      "1,False,num_reads_800genomes,800_genomes_1000000_reads,EnrichSeq_k35m31,800,1000000,0.8044328552803129,0.8533886583679114,2.027331306816969e-06\n",
      "1,False,num_reads_800genomes,800_genomes_1000000_reads,Bracken,800,1000000,0.2985658409387223,0.5518072289156627,8.387732149710782e-07\n",
      "1,False,num_reads_800genomes,800_genomes_200000_reads,EnrichSeq_k75m31,800,200000,0.13820078226857888,0.8760330578512396,0.00012600493996668224\n",
      "1,False,num_reads_800genomes,800_genomes_200000_reads,EnrichSeq_k35m31,800,200000,0.16166883963494133,0.8051948051948052,7.853330216834257e-05\n",
      "1,False,num_reads_800genomes,800_genomes_200000_reads,EnrichSeq_k28m28,800,200000,0.14993481095176012,0.5989583333333334,4.458869998584403e-05\n",
      "1,False,num_reads_800genomes,800_genomes_200000_reads,Bracken,800,200000,0.2920469361147327,0.5517241379310345,1.025992085358192e-06\n",
      "1,False,num_reads_200genomes,200_genomes_500000_reads,EnrichSeq_k75m31,200,500000,0.91,0.8310502283105022,8.03421688632841e-06\n",
      "1,False,num_reads_200genomes,200_genomes_500000_reads,EnrichSeq_k35m31,200,500000,0.92,0.8141592920353983,7.548468691784142e-06\n",
      "1,False,num_reads_200genomes,200_genomes_500000_reads,EnrichSeq_k28m28,200,500000,0.9,0.46875,2.526798803770221e-06\n",
      "1,False,num_reads_200genomes,200_genomes_500000_reads,Bracken,200,500000,0.31,0.35428571428571426,3.412938640963574e-06\n",
      "1,False,num_reads_200genomes,200_genomes_100000_reads,EnrichSeq_k75m31,200,100000,0.35,0.9210526315789473,0.0001534582142766644\n",
      "1,False,num_reads_200genomes,200_genomes_100000_reads,EnrichSeq_k35m31,200,100000,0.375,0.872093023255814,0.00011469659921780617\n",
      "1,False,num_reads_200genomes,200_genomes_100000_reads,EnrichSeq_k28m28,200,100000,0.33,0.6470588235294118,5.3513588406733054e-05\n",
      "1,False,num_reads_200genomes,200_genomes_100000_reads,Bracken,200,100000,0.31,0.33879781420765026,3.5292967600444264e-06\n",
      "1,False,num_reads_200genomes,200_genomes_800000_reads,EnrichSeq_k75m31,200,800000,0.92,0.7896995708154506,7.3534455311774685e-06\n",
      "1,False,num_reads_200genomes,200_genomes_800000_reads,EnrichSeq_k35m31,200,800000,0.94,0.7768595041322314,7.0101332771790834e-06\n",
      "1,False,num_reads_200genomes,200_genomes_800000_reads,EnrichSeq_k28m28,200,800000,0.93,0.46851385390428213,2.4821432134452995e-06\n",
      "1,False,num_reads_200genomes,200_genomes_800000_reads,Bracken,200,800000,0.31,0.3502824858757062,3.3932218149522574e-06\n",
      "1,False,num_reads_200genomes,200_genomes_300000_reads,EnrichSeq_k75m31,200,300000,0.81,0.8901098901098901,1.1844708114665681e-05\n",
      "1,False,num_reads_200genomes,200_genomes_300000_reads,EnrichSeq_k35m31,200,300000,0.815,0.8624338624338624,1.1476194271051955e-05\n",
      "1,False,num_reads_200genomes,200_genomes_300000_reads,EnrichSeq_k28m28,200,300000,0.795,0.5618374558303887,3.24312138474804e-06\n",
      "1,False,num_reads_200genomes,200_genomes_300000_reads,Bracken,200,300000,0.31,0.34444444444444444,3.1964371230325194e-06\n",
      "1,False,num_reads_200genomes,200_genomes_1000000_reads,EnrichSeq_k75m31,200,1000000,0.935,0.8025751072961373,6.9312221811633e-06\n",
      "1,False,num_reads_200genomes,200_genomes_1000000_reads,EnrichSeq_k35m31,200,1000000,0.955,0.7549407114624506,6.330207697851986e-06\n",
      "1,False,num_reads_200genomes,200_genomes_1000000_reads,EnrichSeq_k28m28,200,1000000,0.94,0.4486873508353222,2.8653207056751576e-06\n",
      "1,False,num_reads_200genomes,200_genomes_1000000_reads,Bracken,200,1000000,0.31,0.3502824858757062,3.512698698805654e-06\n",
      "1,False,num_reads_200genomes,200_genomes_600000_reads,EnrichSeq_k75m31,200,600000,0.915,0.8133333333333334,8.010157394819027e-06\n",
      "1,False,num_reads_200genomes,200_genomes_600000_reads,EnrichSeq_k35m31,200,600000,0.935,0.7759336099585062,7.352489422627741e-06\n",
      "1,False,num_reads_200genomes,200_genomes_600000_reads,EnrichSeq_k28m28,200,600000,0.91,0.46547314578005117,2.4390690819344536e-06\n",
      "1,False,num_reads_200genomes,200_genomes_600000_reads,Bracken,200,600000,0.31,0.3463687150837989,3.3167525430054423e-06\n",
      "1,False,num_reads_200genomes,200_genomes_200000_reads,EnrichSeq_k75m31,200,200000,0.7,0.9032258064516129,1.8042906964951144e-05\n",
      "1,False,num_reads_200genomes,200_genomes_200000_reads,EnrichSeq_k35m31,200,200000,0.735,0.8647058823529412,1.565445094063179e-05\n",
      "1,False,num_reads_200genomes,200_genomes_200000_reads,EnrichSeq_k28m28,200,200000,0.685,0.5637860082304527,4.540705747694094e-06\n",
      "1,False,num_reads_200genomes,200_genomes_200000_reads,Bracken,200,200000,0.31,0.34444444444444444,3.2319408052367292e-06\n",
      "1,False,num_reads_200genomes,200_genomes_700000_reads,EnrichSeq_k75m31,200,700000,0.935,0.8095238095238095,7.189611892782473e-06\n",
      "1,False,num_reads_200genomes,200_genomes_700000_reads,EnrichSeq_k35m31,200,700000,0.945,0.7590361445783133,6.54967530292645e-06\n",
      "1,False,num_reads_200genomes,200_genomes_700000_reads,EnrichSeq_k28m28,200,700000,0.92,0.44878048780487806,2.511016766978701e-06\n",
      "1,False,num_reads_200genomes,200_genomes_700000_reads,Bracken,200,700000,0.31,0.3502824858757062,3.4945945677997095e-06\n",
      "1,False,num_reads_200genomes,200_genomes_400000_reads,EnrichSeq_k75m31,200,400000,0.85,0.8854166666666666,9.762069829389916e-06\n",
      "1,False,num_reads_200genomes,200_genomes_400000_reads,EnrichSeq_k35m31,200,400000,0.885,0.8388625592417062,9.291374899924677e-06\n",
      "1,False,num_reads_200genomes,200_genomes_400000_reads,EnrichSeq_k28m28,200,400000,0.865,0.5195195195195195,2.9165109275819467e-06\n",
      "1,False,num_reads_200genomes,200_genomes_400000_reads,Bracken,200,400000,0.31,0.34444444444444444,3.6424966993999896e-06\n",
      "1,False,num_reads_200genomes,200_genomes_900000_reads,EnrichSeq_k75m31,200,900000,0.955,0.809322033898305,6.886234592437491e-06\n",
      "1,False,num_reads_200genomes,200_genomes_900000_reads,EnrichSeq_k35m31,200,900000,0.96,0.7559055118110236,6.501462849342177e-06\n",
      "1,False,num_reads_200genomes,200_genomes_900000_reads,EnrichSeq_k28m28,200,900000,0.93,0.4376470588235294,2.900063561994799e-06\n",
      "1,False,num_reads_200genomes,200_genomes_900000_reads,Bracken,200,900000,0.31,0.34831460674157305,3.4699767447479468e-06\n",
      "number not found: 1\n",
      "number not found: 1\n",
      "number not found: 1\n",
      "number not found: 1\n",
      "number not found: 1\n",
      "number not found: 1\n",
      "number not found: 1\n",
      "number not found: 1\n",
      "number not found: 1\n",
      "number not found: 1\n",
      "number not found: 1\n",
      "number not found: 1\n",
      "number not found: 1\n",
      "ERROR:\n",
      "\tFVE: NC_029072.1 NOT CONVERTED\n",
      "\tabundances: 1.00499\n",
      "ERROR:\n",
      "\tFVE: NC_029069.1 NOT CONVERTED\n",
      "\tabundances: 2.26741\n",
      "ERROR:\n",
      "\tFVE: NC_029069.1 NOT CONVERTED\n",
      "\tabundances: 1.16361\n",
      "ERROR:\n",
      "\tFVE: NC_029072.1 NOT CONVERTED\n",
      "\tabundances: 1.01201\n",
      "ERROR:\n",
      "\tFVE: NC_029072.1 NOT CONVERTED\n",
      "\tabundances: 1.01201\n",
      "EnrichSeq_k28m28 is missing 800_genomes_1000000_reads from num_reads_800genomes\n",
      "EnrichSeq_k28m28 is missing 800_genomes_800000_reads from num_reads_800genomes\n",
      "EnrichSeq_k28m28 is missing 800_genomes_900000_reads from num_reads_800genomes\n",
      "2,False,num_genomes,800_genomes_500000_reads,FastViromeExplorer,800,500000,0.4870801033591731,0.843400447427293,1.1504012838551443e-06\n",
      "2,False,num_genomes,800_genomes_500000_reads,EnrichSeq_k75m31,800,500000,0.5478036175710594,0.9118279569892473,5.43990825085499e-06\n",
      "2,False,num_genomes,800_genomes_500000_reads,EnrichSeq_k35m31,800,500000,0.5813953488372093,0.9018036072144289,4.67555683186843e-06\n",
      "2,False,num_genomes,800_genomes_500000_reads,EnrichSeq_k28m28,800,500000,0.537467700258398,0.6864686468646864,1.7750198995529908e-06\n",
      "2,False,num_genomes,800_genomes_500000_reads,Bracken,800,500000,0.2958656330749354,0.531322505800464,7.794652315515306e-07\n",
      "2,False,num_genomes,200_genomes_500000_reads,FastViromeExplorer,200,500000,0.53,0.5760869565217391,3.5186314728166676e-06\n",
      "2,False,num_genomes,200_genomes_500000_reads,EnrichSeq_k75m31,200,500000,0.915,0.8394495412844036,5.644191324627986e-06\n",
      "2,False,num_genomes,200_genomes_500000_reads,EnrichSeq_k35m31,200,500000,0.925,0.8008658008658008,5.381820773564675e-06\n",
      "2,False,num_genomes,200_genomes_500000_reads,EnrichSeq_k28m28,200,500000,0.93,0.5723076923076923,3.5389289976833063e-06\n",
      "2,False,num_genomes,200_genomes_500000_reads,Bracken,200,500000,0.325,0.35911602209944754,2.711557684288955e-06\n",
      "2,False,num_genomes,400_genomes_500000_reads,FastViromeExplorer,400,500000,0.4475,0.6302816901408451,2.3026953217878023e-06\n",
      "2,False,num_genomes,400_genomes_500000_reads,EnrichSeq_k75m31,400,500000,0.755,0.880466472303207,3.966630049085686e-06\n",
      "2,False,num_genomes,400_genomes_500000_reads,EnrichSeq_k35m31,400,500000,0.7775,0.8474114441416893,3.6505921519937618e-06\n",
      "2,False,num_genomes,400_genomes_500000_reads,EnrichSeq_k28m28,400,500000,0.7475,0.5526802218114603,1.457530273174146e-06\n",
      "2,False,num_genomes,400_genomes_500000_reads,Bracken,400,500000,0.28,0.4043321299638989,1.664338290498958e-06\n",
      "2,False,num_genomes,1000_genomes_500000_reads,FastViromeExplorer,1000,500000,0.46737967914438505,0.8468992248062015,9.457385261375829e-07\n",
      "2,False,num_genomes,1000_genomes_500000_reads,EnrichSeq_k75m31,1000,500000,0.4320855614973262,0.926605504587156,7.095248144116796e-06\n",
      "2,False,num_genomes,1000_genomes_500000_reads,EnrichSeq_k35m31,1000,500000,0.4802139037433155,0.905241935483871,5.686894621931209e-06\n",
      "2,False,num_genomes,1000_genomes_500000_reads,EnrichSeq_k28m28,1000,500000,0.45454545454545453,0.6933115823817292,2.0170836927088567e-06\n",
      "2,False,num_genomes,1000_genomes_500000_reads,Bracken,1000,500000,0.28342245989304815,0.5342741935483871,6.998430676799225e-07\n",
      "2,False,num_genomes,600_genomes_500000_reads,FastViromeExplorer,600,500000,0.4741235392320534,0.7780821917808219,1.5954466359640254e-06\n",
      "2,False,num_genomes,600_genomes_500000_reads,EnrichSeq_k75m31,600,500000,0.6427378964941569,0.9390243902439024,4.673576784337125e-06\n",
      "2,False,num_genomes,600_genomes_500000_reads,EnrichSeq_k35m31,600,500000,0.676126878130217,0.8980044345898004,3.928502367648689e-06\n",
      "2,False,num_genomes,600_genomes_500000_reads,EnrichSeq_k28m28,600,500000,0.6527545909849749,0.6683760683760683,1.602136582639032e-06\n",
      "2,False,num_genomes,600_genomes_500000_reads,Bracken,600,500000,0.2838063439065108,0.47619047619047616,9.702708641783411e-07\n",
      "2,False,num_genomes,10_genomes_500000_reads,FastViromeExplorer,10,500000,0.5,0.16666666666666666,4.374601962276535e-05\n",
      "2,False,num_genomes,10_genomes_500000_reads,EnrichSeq_k75m31,10,500000,0.8,0.8888888888888888,0.0004992170083494314\n",
      "2,False,num_genomes,10_genomes_500000_reads,EnrichSeq_k28m28,10,500000,0.8,0.6666666666666666,0.0005920432643704548\n",
      "2,False,num_genomes,10_genomes_500000_reads,Bracken,10,500000,0.1,0.03225806451612903,1.3587179913389417e-05\n",
      "2,False,num_reads_800genomes,800_genomes_500000_reads,FastViromeExplorer,800,500000,0.4870801033591731,0.8510158013544018,1.187108747476771e-06\n",
      "2,False,num_reads_800genomes,800_genomes_500000_reads,EnrichSeq_k75m31,800,500000,0.5271317829457365,0.9272727272727272,6.1545226492287374e-06\n",
      "2,False,num_reads_800genomes,800_genomes_500000_reads,EnrichSeq_k35m31,800,500000,0.5723514211886305,0.8985801217038539,4.943263429147697e-06\n",
      "2,False,num_reads_800genomes,800_genomes_500000_reads,EnrichSeq_k28m28,800,500000,0.5568475452196382,0.6985413290113452,1.6543399595667015e-06\n",
      "2,False,num_reads_800genomes,800_genomes_500000_reads,Bracken,800,500000,0.2958656330749354,0.531322505800464,8.232626425717777e-07\n",
      "2,False,num_reads_800genomes,800_genomes_400000_reads,FastViromeExplorer,800,400000,0.4870801033591731,0.8471910112359551,1.1735714843105642e-06\n",
      "2,False,num_reads_800genomes,800_genomes_400000_reads,EnrichSeq_k75m31,800,400000,0.41214470284237725,0.9011299435028248,1.0250664292079996e-05\n",
      "2,False,num_reads_800genomes,800_genomes_400000_reads,EnrichSeq_k35m31,800,400000,0.4483204134366925,0.8784810126582279,8.608632238643657e-06\n",
      "2,False,num_reads_800genomes,800_genomes_400000_reads,EnrichSeq_k28m28,800,400000,0.4315245478036176,0.6774847870182555,2.847453782982687e-06\n",
      "2,False,num_reads_800genomes,800_genomes_400000_reads,Bracken,800,400000,0.29457364341085274,0.5339578454332553,8.739319466354739e-07\n",
      "2,False,num_reads_800genomes,800_genomes_300000_reads,FastViromeExplorer,800,300000,0.4870801033591731,0.843400447427293,1.146604318208889e-06\n",
      "2,False,num_reads_800genomes,800_genomes_300000_reads,EnrichSeq_k75m31,800,300000,0.27260981912144705,0.9094827586206896,2.9388479443525913e-05\n",
      "2,False,num_reads_800genomes,800_genomes_300000_reads,EnrichSeq_k35m31,800,300000,0.3152454780361757,0.856140350877193,2.030769016849996e-05\n",
      "2,False,num_reads_800genomes,800_genomes_300000_reads,EnrichSeq_k28m28,800,300000,0.2751937984496124,0.6870967741935484,8.317172950708808e-06\n",
      "2,False,num_reads_800genomes,800_genomes_300000_reads,Bracken,800,300000,0.29069767441860467,0.5244755244755245,7.54016111238657e-07\n",
      "2,False,num_reads_800genomes,800_genomes_100000_reads,FastViromeExplorer,800,100000,0.48320413436692505,0.8480725623582767,1.1930045910936402e-06\n",
      "2,False,num_reads_800genomes,800_genomes_100000_reads,EnrichSeq_k75m31,800,100000,0.04263565891472868,0.8461538461538461,0.0005779752391540247\n",
      "2,False,num_reads_800genomes,800_genomes_100000_reads,EnrichSeq_k35m31,800,100000,0.056847545219638244,0.8148148148148148,0.0006542971562645843\n",
      "2,False,num_reads_800genomes,800_genomes_100000_reads,EnrichSeq_k28m28,800,100000,0.04263565891472868,0.6226415094339622,0.0002662510951003948\n",
      "2,False,num_reads_800genomes,800_genomes_100000_reads,Bracken,800,100000,0.28940568475452194,0.5209302325581395,1.035922589595348e-06\n",
      "2,False,num_reads_800genomes,800_genomes_900000_reads,FastViromeExplorer,800,900000,0.4883720930232558,0.8513513513513513,1.1757726877320306e-06\n",
      "2,False,num_reads_800genomes,800_genomes_900000_reads,EnrichSeq_k75m31,800,900000,0.7764857881136951,0.9051204819277109,1.6282455643936448e-06\n",
      "2,False,num_reads_800genomes,800_genomes_900000_reads,EnrichSeq_k35m31,800,900000,0.7984496124031008,0.8803418803418803,1.8089220931690599e-06\n",
      "2,False,num_reads_800genomes,800_genomes_900000_reads,Bracken,800,900000,0.2971576227390181,0.5361305361305362,7.797163116369738e-07\n",
      "2,False,num_reads_800genomes,800_genomes_600000_reads,FastViromeExplorer,800,600000,0.4883720930232558,0.8475336322869955,1.1501870900008159e-06\n",
      "2,False,num_reads_800genomes,800_genomes_600000_reads,EnrichSeq_k75m31,800,600000,0.6395348837209303,0.9049360146252285,3.5659723575109563e-06\n",
      "2,False,num_reads_800genomes,800_genomes_600000_reads,EnrichSeq_k35m31,800,600000,0.6640826873385013,0.8923611111111112,3.397297130093733e-06\n",
      "2,False,num_reads_800genomes,800_genomes_600000_reads,EnrichSeq_k28m28,800,600000,0.6369509043927648,0.6895104895104895,1.1195141976267078e-06\n",
      "2,False,num_reads_800genomes,800_genomes_600000_reads,Bracken,800,600000,0.29844961240310075,0.5298165137614679,7.945930325549141e-07\n",
      "2,False,num_reads_800genomes,800_genomes_800000_reads,FastViromeExplorer,800,800000,0.4883720930232558,0.8532731376975169,1.1771384580999388e-06\n",
      "2,False,num_reads_800genomes,800_genomes_800000_reads,EnrichSeq_k75m31,800,800000,0.7390180878552972,0.9137380191693291,2.3094018614046413e-06\n",
      "2,False,num_reads_800genomes,800_genomes_800000_reads,EnrichSeq_k35m31,800,800000,0.7674418604651163,0.8709677419354839,1.9209122894776733e-06\n",
      "2,False,num_reads_800genomes,800_genomes_800000_reads,Bracken,800,800000,0.29844961240310075,0.528604118993135,7.246284990896863e-07\n",
      "2,False,num_reads_800genomes,800_genomes_700000_reads,FastViromeExplorer,800,700000,0.4870801033591731,0.8452914798206278,1.1641484764634565e-06\n",
      "2,False,num_reads_800genomes,800_genomes_700000_reads,EnrichSeq_k75m31,800,700000,0.7002583979328165,0.9124579124579124,2.5281213742181095e-06\n",
      "2,False,num_reads_800genomes,800_genomes_700000_reads,EnrichSeq_k35m31,800,700000,0.7209302325581395,0.8885350318471338,2.6048260089159407e-06\n",
      "2,False,num_reads_800genomes,800_genomes_700000_reads,EnrichSeq_k28m28,800,700000,0.7015503875968992,0.6510791366906474,8.36427169598904e-07\n",
      "2,False,num_reads_800genomes,800_genomes_700000_reads,Bracken,800,700000,0.29844961240310075,0.5334872979214781,7.868238208439669e-07\n",
      "2,False,num_reads_800genomes,800_genomes_1000000_reads,FastViromeExplorer,800,1000000,0.4883720930232558,0.8513513513513513,1.1756009533591407e-06\n",
      "2,False,num_reads_800genomes,800_genomes_1000000_reads,EnrichSeq_k75m31,800,1000000,0.7997416020671835,0.8792613636363636,1.6933873434675534e-06\n",
      "2,False,num_reads_800genomes,800_genomes_1000000_reads,EnrichSeq_k35m31,800,1000000,0.8229974160206718,0.8798342541436464,1.6554076501325302e-06\n",
      "2,False,num_reads_800genomes,800_genomes_1000000_reads,Bracken,800,1000000,0.29844961240310075,0.5298165137614679,7.232664479775399e-07\n",
      "2,False,num_reads_800genomes,800_genomes_200000_reads,FastViromeExplorer,800,200000,0.4870801033591731,0.8452914798206278,1.1432848190580135e-06\n",
      "2,False,num_reads_800genomes,800_genomes_200000_reads,EnrichSeq_k75m31,800,200000,0.13565891472868216,0.9722222222222222,0.00015619750117559021\n",
      "2,False,num_reads_800genomes,800_genomes_200000_reads,EnrichSeq_k35m31,800,200000,0.16020671834625322,0.8794326241134752,7.736605699389748e-05\n",
      "2,False,num_reads_800genomes,800_genomes_200000_reads,EnrichSeq_k28m28,800,200000,0.12790697674418605,0.6971830985915493,3.8825344023389097e-05\n",
      "2,False,num_reads_800genomes,800_genomes_200000_reads,Bracken,800,200000,0.29457364341085274,0.5290023201856149,9.159660814042021e-07\n",
      "2,False,num_reads_200genomes,200_genomes_500000_reads,FastViromeExplorer,200,500000,0.53,0.5921787709497207,3.7560795224045255e-06\n",
      "2,False,num_reads_200genomes,200_genomes_500000_reads,EnrichSeq_k75m31,200,500000,0.92,0.8214285714285714,5.3819308487400164e-06\n",
      "2,False,num_reads_200genomes,200_genomes_500000_reads,EnrichSeq_k35m31,200,500000,0.925,0.7644628099173554,5.019690259206916e-06\n",
      "2,False,num_reads_200genomes,200_genomes_500000_reads,EnrichSeq_k28m28,200,500000,0.935,0.520891364902507,2.182920490184563e-06\n",
      "2,False,num_reads_200genomes,200_genomes_500000_reads,Bracken,200,500000,0.325,0.3651685393258427,2.727742717447166e-06\n",
      "2,False,num_reads_200genomes,200_genomes_100000_reads,FastViromeExplorer,200,100000,0.53,0.5824175824175825,3.5161444986477802e-06\n",
      "2,False,num_reads_200genomes,200_genomes_100000_reads,EnrichSeq_k75m31,200,100000,0.405,0.9310344827586207,7.82821681896053e-05\n",
      "2,False,num_reads_200genomes,200_genomes_100000_reads,EnrichSeq_k35m31,200,100000,0.465,0.8942307692307693,5.479479117598885e-05\n",
      "2,False,num_reads_200genomes,200_genomes_100000_reads,EnrichSeq_k28m28,200,100000,0.42,0.7058823529411765,2.2600858485984175e-05\n",
      "2,False,num_reads_200genomes,200_genomes_100000_reads,Bracken,200,100000,0.325,0.35135135135135137,3.0726456101400472e-06\n",
      "2,False,num_reads_200genomes,200_genomes_800000_reads,FastViromeExplorer,200,800000,0.53,0.5824175824175825,3.612669419801596e-06\n",
      "2,False,num_reads_200genomes,200_genomes_800000_reads,EnrichSeq_k75m31,200,800000,0.945,0.84375,5.50495201645078e-06\n",
      "2,False,num_reads_200genomes,200_genomes_800000_reads,EnrichSeq_k35m31,200,800000,0.945,0.8111587982832618,5.246855941365286e-06\n",
      "2,False,num_reads_200genomes,200_genomes_800000_reads,EnrichSeq_k28m28,200,800000,0.945,0.4883720930232558,2.124642512335857e-06\n",
      "2,False,num_reads_200genomes,200_genomes_800000_reads,Bracken,200,800000,0.325,0.3611111111111111,2.7410756924066637e-06\n",
      "2,False,num_reads_200genomes,200_genomes_300000_reads,FastViromeExplorer,200,300000,0.53,0.585635359116022,3.6814577051513796e-06\n",
      "2,False,num_reads_200genomes,200_genomes_300000_reads,EnrichSeq_k75m31,200,300000,0.865,0.9057591623036649,7.854926087104871e-06\n",
      "2,False,num_reads_200genomes,200_genomes_300000_reads,EnrichSeq_k35m31,200,300000,0.87,0.8529411764705882,7.021407383670777e-06\n",
      "2,False,num_reads_200genomes,200_genomes_300000_reads,EnrichSeq_k28m28,200,300000,0.845,0.5971731448763251,2.8137246435769645e-06\n",
      "2,False,num_reads_200genomes,200_genomes_300000_reads,Bracken,200,300000,0.325,0.3693181818181818,2.736561026693941e-06\n",
      "2,False,num_reads_200genomes,200_genomes_1000000_reads,FastViromeExplorer,200,1000000,0.53,0.5792349726775956,3.553806748018249e-06\n",
      "2,False,num_reads_200genomes,200_genomes_1000000_reads,EnrichSeq_k75m31,200,1000000,0.945,0.84,5.2615745560162175e-06\n",
      "2,False,num_reads_200genomes,200_genomes_1000000_reads,EnrichSeq_k35m31,200,1000000,0.945,0.8111587982832618,5.099921589099881e-06\n",
      "2,False,num_reads_200genomes,200_genomes_1000000_reads,EnrichSeq_k28m28,200,1000000,0.935,0.4794871794871795,1.9387713103768347e-06\n",
      "2,False,num_reads_200genomes,200_genomes_1000000_reads,Bracken,200,1000000,0.325,0.3693181818181818,2.7074049438233185e-06\n",
      "2,False,num_reads_200genomes,200_genomes_600000_reads,FastViromeExplorer,200,600000,0.53,0.572972972972973,3.487022060626755e-06\n",
      "2,False,num_reads_200genomes,200_genomes_600000_reads,EnrichSeq_k75m31,200,600000,0.94,0.8663594470046083,5.454960203600731e-06\n",
      "2,False,num_reads_200genomes,200_genomes_600000_reads,EnrichSeq_k35m31,200,600000,0.95,0.8016877637130801,4.977309632948323e-06\n",
      "2,False,num_reads_200genomes,200_genomes_600000_reads,EnrichSeq_k28m28,200,600000,0.935,0.5137362637362637,1.953253749275465e-06\n",
      "2,False,num_reads_200genomes,200_genomes_600000_reads,Bracken,200,600000,0.325,0.36312849162011174,2.5996961840498195e-06\n",
      "2,False,num_reads_200genomes,200_genomes_200000_reads,FastViromeExplorer,200,200000,0.53,0.5955056179775281,3.819106783437982e-06\n",
      "2,False,num_reads_200genomes,200_genomes_200000_reads,EnrichSeq_k75m31,200,200000,0.765,0.9053254437869822,1.1409520662689582e-05\n",
      "2,False,num_reads_200genomes,200_genomes_200000_reads,EnrichSeq_k35m31,200,200000,0.785,0.8920454545454546,9.903606918712092e-06\n",
      "2,False,num_reads_200genomes,200_genomes_200000_reads,EnrichSeq_k28m28,200,200000,0.76,0.658008658008658,3.827900093584071e-06\n",
      "2,False,num_reads_200genomes,200_genomes_200000_reads,Bracken,200,200000,0.325,0.36312849162011174,2.9596171409108187e-06\n",
      "2,False,num_reads_200genomes,200_genomes_700000_reads,FastViromeExplorer,200,700000,0.53,0.5792349726775956,3.583076665952866e-06\n",
      "2,False,num_reads_200genomes,200_genomes_700000_reads,EnrichSeq_k75m31,200,700000,0.945,0.8590909090909091,5.312414834551195e-06\n",
      "2,False,num_reads_200genomes,200_genomes_700000_reads,EnrichSeq_k35m31,200,700000,0.95,0.811965811965812,5.174184572262021e-06\n",
      "2,False,num_reads_200genomes,200_genomes_700000_reads,EnrichSeq_k28m28,200,700000,0.94,0.47837150127226463,2.0495625347609173e-06\n",
      "2,False,num_reads_200genomes,200_genomes_700000_reads,Bracken,200,700000,0.325,0.3672316384180791,2.6844720027043153e-06\n",
      "2,False,num_reads_200genomes,200_genomes_400000_reads,FastViromeExplorer,200,400000,0.53,0.5668449197860963,3.4428309466839e-06\n",
      "2,False,num_reads_200genomes,200_genomes_400000_reads,EnrichSeq_k75m31,200,400000,0.89,0.898989898989899,6.620000056175127e-06\n",
      "2,False,num_reads_200genomes,200_genomes_400000_reads,EnrichSeq_k35m31,200,400000,0.895,0.8564593301435407,6.589855908797804e-06\n",
      "2,False,num_reads_200genomes,200_genomes_400000_reads,EnrichSeq_k28m28,200,400000,0.9,0.5769230769230769,2.428350021667876e-06\n",
      "2,False,num_reads_200genomes,200_genomes_400000_reads,Bracken,200,400000,0.32,0.3699421965317919,2.9860595698741324e-06\n",
      "2,False,num_reads_200genomes,200_genomes_900000_reads,FastViromeExplorer,200,900000,0.53,0.5824175824175825,3.6392526481839694e-06\n",
      "2,False,num_reads_200genomes,200_genomes_900000_reads,EnrichSeq_k75m31,200,900000,0.945,0.8709677419354839,5.420796576283313e-06\n",
      "2,False,num_reads_200genomes,200_genomes_900000_reads,EnrichSeq_k35m31,200,900000,0.945,0.7875,4.8943336843303e-06\n",
      "2,False,num_reads_200genomes,200_genomes_900000_reads,EnrichSeq_k28m28,200,900000,0.935,0.4973404255319149,2.0374140669028623e-06\n",
      "2,False,num_reads_200genomes,200_genomes_900000_reads,Bracken,200,900000,0.325,0.3651685393258427,2.619824759326909e-06\n"
     ]
    }
   ],
   "source": [
    "benchmarking_directories = [\n",
    "                             [\"//scratch/summit/dral3008/benchmarking_enrichseq/RESULTS/results-NEW/test_1/\",\n",
    "                              \"//scratch/summit/dral3008/benchmarking_enrichseq/TESTS/tests_new/test_1/\"],\n",
    "                             [\"//scratch/summit/dral3008/benchmarking_enrichseq/RESULTS/results-NEW/test_2/\",\n",
    "                              \"//scratch/summit/dral3008/benchmarking_enrichseq/TESTS/tests_new/test_2/\"],\n",
    "                             [\"//scratch/summit/dral3008/benchmarking_enrichseq/RESULTS/results-NEW/test_3/\",\n",
    "                              \"//scratch/summit/dral3008/benchmarking_enrichseq/TESTS/tests_new/test_3/\"],\n",
    "                            ]\n",
    "\n",
    "def create_results_csv(benchmarking_directories=benchmarking_directories):\n",
    "    \"\"\"\n",
    "    this method prints out the results\n",
    "    \"\"\"\n",
    "    with open(\"results_file.csv\", 'a') as results_file:\n",
    "        print(f\"replicate_ID,MergeOverlap,major_test,sub_test,tool,genome_count,read_count,classification_recall,classification_precision,l2_abundance_distance\")\n",
    "        results_file.write(f\"replicate_ID,MergeOverlap,major_test,sub_test,tool,genome_count,read_count,classification_recall,classification_precision,l2_abundance_distance\\n\")\n",
    "        for iteration, test_results in enumerate(benchmarking_directories):\n",
    "            results = parse_dir(test_results[0], test_results[1])\n",
    "            for major_test in results.keys(): # ['num_genomes', 'config_files', 'num_reads_800genomes', 'num_reads_200genomes']\n",
    "                for sub_test in results[major_test].keys():\n",
    "                    for tool, result in results[major_test][sub_test].items():\n",
    "                        genomes = sub_test.split(\"genomes\")[0].strip(\"_\")\n",
    "                        reads = sub_test.split(\"reads\")[0].split(\"_\")[-2]\n",
    "                        print(f\"{iteration},False,{major_test},{sub_test},{tool},{genomes},{reads},{result.classification_recall},{result.classification_precision},{np.average(result.l2_abundance_distance)}\")\n",
    "                        results_file.write(f\"{iteration},False,{major_test},{sub_test},{tool},{genomes},{reads},{result.classification_recall},{result.classification_precision},{np.average(result.l2_abundance_distance)}\\n\")\n",
    "\n",
    "create_results_csv()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plotting Methods\n",
    "These methods use the common `ResultStruct` datastructure (or arry thereof) for plotting individual tool metrics. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_barplot(input_results, test, metric='classification'):                                                        \n",
    "    \"\"\"                                                                         \n",
    "    This function plots a bar plot output for a given test set and value\n",
    "    \"\"\" \n",
    "    \n",
    "    width = 0.35\n",
    "    fig = plt.figure()\n",
    "    ax = fig.add_axes([0,0,1,1])\n",
    "    x_tick_names = [[], [], [], []]\n",
    "    index_map = {'EnrichSeq': 0, 'EnrichSeq_wMergeOverlap' : 1, 'FastViromeExplorer' : 2, 'Bracken' : 3}\n",
    "    if (metric == 'classification'):\n",
    "        sub_results = input_results[test]\n",
    "        x_tick_arr = []\n",
    "        for subtest_name in sub_results.keys():\n",
    "            x_tick_arr.append(\" \".join(subtest_name.split(\"_\")))\n",
    "            print(subtest_name)\n",
    "            for tool, results_struct in sub_results[subtest_name].items():\n",
    "                print(tool)\n",
    "                print(f\"Recall: {results_struct.classification_recall}\")\n",
    "                print(f\"Precision: {results_struct.classification_precision}\")\n",
    "                f1_score = 2 * (results_struct.classification_recall * results_struct.classification_precision) \n",
    "                if ((results_struct.classification_recall != 0) or\n",
    "                    (results_struct.classification_precision != 0)):\n",
    "                    f1_score /= (results_struct.classification_recall + results_struct.classification_precision) \n",
    "                    f1_score = results_struct.classification_recall\n",
    "                x_tick_names[index_map[tool]].append(f1_score)\n",
    "        \n",
    "    elif (metric == 'abundance'):\n",
    "        sub_results = input_results[test]\n",
    "        x_tick_arr = []\n",
    "        for subtest_name in sub_results.keys():\n",
    "            x_tick_arr.append(\" \".join(subtest_name.split(\"_\")))\n",
    "            for tool, results_struct in sub_results[subtest_name].items():\n",
    "                l2_avg = np.average(results_struct.l2_abundance_distance)\n",
    "                x_tick_names[index_map[tool]].append(l2_avg)\n",
    "    else:\n",
    "        print(f\"Metric of type {metric} is not known\")\n",
    "        \n",
    "        \n",
    "    # make plot\n",
    "    ind = np.arange(len(x_tick_names[0]))\n",
    "    ax.bar(ind, x_tick_names[index_map['EnrichSeq']], color='purple', width = 0.20)\n",
    "    if (len(x_tick_names[0]) == len(x_tick_names[1])):\n",
    "        ax.bar(ind + 0.20, x_tick_names[index_map['EnrichSeq_wMergeOverlap']], color='r', width = 0.20)\n",
    "        ax.bar(ind + 0.40, x_tick_names[index_map['FastViromeExplorer']], width = 0.20, color='b')\n",
    "        ax.bar(ind + 0.60, x_tick_names[index_map['Bracken']], width = 0.20, color='g')\n",
    "    ax.legend(labels=['EnrichSeq', 'EnrichSeq_wMergeOverlap', 'FastViromeExplorer', 'Bracken'])\n",
    "    ax.set_xticks(ind + 0.25)\n",
    "    ax.set_xticklabels(x_tick_arr, rotation=90)\n",
    "    ax.set_title(f\"Testing the impact of {test} \\n {metric}-focused\", size=15)\n",
    "    ax.set_ylabel(\"Recall\", size=15)\n",
    "    ax.set_xlabel(\"Sub test\", size=15)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Benchmarking\n",
    "This section splits the various benchmarking sections into a clear, concise set of scripts."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Number of genomes comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "800_genomes_500000_reads\n",
      "FastViromeExplorer\n",
      "Recall: 0.4774193548387097\n",
      "Precision: 0.8333333333333334\n",
      "EnrichSeq_k75m31\n",
      "Recall: 0.7148387096774194\n",
      "Precision: 0.8835725677830941\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'EnrichSeq_k75m31'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-31-f60bec4a0b24>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mplot_barplot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresults\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtest\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'num_genomes'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmetric\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'classification'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mplot_barplot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresults\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtest\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'num_genomes'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmetric\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'abundance'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mresults\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'num_genomes'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-30-f76e03576754>\u001b[0m in \u001b[0;36mplot_barplot\u001b[0;34m(input_results, test, metric)\u001b[0m\n\u001b[1;32m     24\u001b[0m                     \u001b[0mf1_score\u001b[0m \u001b[0;34m/=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mresults_struct\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclassification_recall\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mresults_struct\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclassification_precision\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     25\u001b[0m                     \u001b[0mf1_score\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mresults_struct\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclassification_recall\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 26\u001b[0;31m                 \u001b[0mx_tick_names\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mindex_map\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mtool\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf1_score\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     27\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     28\u001b[0m     \u001b[0;32melif\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mmetric\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'abundance'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'EnrichSeq_k75m31'"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAd0AAAFDCAYAAAB/UdRdAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjMsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+AADFEAAAPTklEQVR4nO3dUYil91nH8d/TxCi2tRWzhZJNTMStdQlC6xAqglZaJcnF5qZKAkVbQhfUKGgRIkqVeGWLCEK0XbFUBZumXugiK7nQlIqYki3V0KQE1libIULWWnNT2hh9vJhRptPZzJvtOc/2TD4fGDjvOf858/BnmG/Oe86+qe4OALB+r7jSAwDAy4XoAsAQ0QWAIaILAENEFwCGiC4ADDk0ulX14ap6tqo+e4nHq6p+r6ouVNVjVfXm1Y8JAJtvySvdjyS59UUevy3Jid2v00n+4BsfCwCOnkOj292fTPIfL7LkjiR/0jseSfLaqnr9qgYEgKNiFe/pXpfk6T3H27v3AQB7XL2C56gD7jvw2pJVdTo7p6Dzyle+8gff+MY3ruDHA8CcT3/60//e3ccu53tXEd3tJNfvOT6e5JmDFnb3mSRnkmRra6vPnz+/gh8PAHOq6l8v93tXcXr5bJKf3v0U81uSPNfd/7aC5wWAI+XQV7pV9dEkb01ybVVtJ/mNJN+SJN39wSTnktye5EKSLyd597qGBYBNdmh0u/uuQx7vJD+/sokA4IhyRSoAGCK6ADBEdAFgiOgCwBDRBYAhogsAQ0QXAIaILgAMEV0AGCK6ADBEdAFgiOgCwBDRBYAhogsAQ0QXAIaILgAMEV0AGCK6ADBEdAFgiOgCwBDRBYAhogsAQ0QXAIaILgAMEV0AGCK6ADBEdAFgiOgCwBDRBYAhogsAQ0QXAIaILgAMEV0AGCK6ADBEdAFgiOgCwBDRBYAhogsAQ0QXAIaILgAMEV0AGCK6ADBEdAFgiOgCwBDRBYAhogsAQ0QXAIaILgAMEV0AGLIoulV1a1U9WVUXqureAx6/oaoerqrPVNVjVXX76kcFgM12aHSr6qok9ye5LcnJJHdV1cl9y349yYPd/aYkdyb5/VUPCgCbbskr3VuSXOjup7r7+SQPJLlj35pO8h27t1+T5JnVjQgAR8OS6F6X5Ok9x9u79+31m0neWVXbSc4l+YWDnqiqTlfV+ao6f/HixcsYFwA215Lo1gH39b7ju5J8pLuPJ7k9yZ9W1dc9d3ef6e6t7t46duzYS58WADbYkuhuJ7l+z/HxfP3p47uTPJgk3f0PSb4tybWrGBAAjool0X00yYmquqmqrsnOB6XO7lvzhSRvS5Kq+v7sRNf5YwDY49DodvcLSe5J8lCSz2XnU8qPV9V9VXVqd9l7k7ynqv4pyUeTvKu795+CBoCXtauXLOruc9n5gNTe+9635/YTSX54taMBwNHiilQAMER0AWCI6ALAENEFgCGiCwBDRBcAhoguAAwRXQAYIroAMER0AWCI6ALAENEFgCGiCwBDRBcAhoguAAwRXQAYIroAMER0AWCI6ALAENEFgCGiCwBDRBcAhoguAAwRXQAYIroAMER0AWCI6ALAENEFgCGiCwBDRBcAhoguAAwRXQAYIroAMER0AWCI6ALAENEFgCGiCwBDRBcAhoguAAwRXQAYIroAMER0AWCI6ALAENEFgCGiCwBDRBcAhoguAAwRXQAYsii6VXVrVT1ZVReq6t5LrPmpqnqiqh6vqj9b7ZgAsPmuPmxBVV2V5P4kP55kO8mjVXW2u5/Ys+ZEkl9N8sPd/aWqet26BgaATbXkle4tSS5091Pd/XySB5LcsW/Ne5Lc391fSpLufna1YwLA5lsS3euSPL3neHv3vr3ekOQNVfX3VfVIVd26qgEB4Kg49PRykjrgvj7geU4keWuS40n+rqpu7u7//Jonqjqd5HSS3HDDDS95WADYZEte6W4nuX7P8fEkzxyw5i+7+7+6+1+SPJmdCH+N7j7T3VvdvXXs2LHLnRkANtKS6D6a5ERV3VRV1yS5M8nZfWv+IsmPJUlVXZud081PrXJQANh0h0a3u19Ick+Sh5J8LsmD3f14Vd1XVad2lz2U5ItV9USSh5P8Snd/cV1DA8Amqu79b8/O2Nra6vPnz1+Rnw0Al6uqPt3dW5fzva5IBQBDRBcAhoguAAwRXQAYIroAMER0AWCI6ALAENEFgCGiCwBDRBcAhoguAAwRXQAYIroAMER0AWCI6ALAENEFgCGiCwBDRBcAhoguAAwRXQAYIroAMER0AWCI6ALAENEFgCGiCwBDRBcAhoguAAwRXQAYIroAMER0AWCI6ALAENEFgCGiCwBDRBcAhoguAAwRXQAYIroAMER0AWCI6ALAENEFgCGiCwBDRBcAhoguAAwRXQAYIroAMER0AWCI6ALAENEFgCGiCwBDFkW3qm6tqier6kJV3fsi695RVV1VW6sbEQCOhkOjW1VXJbk/yW1JTia5q6pOHrDu1Ul+McmnVj0kABwFS17p3pLkQnc/1d3PJ3kgyR0HrPutJO9P8pUVzgcAR8aS6F6X5Ok9x9u79/2/qnpTkuu7+69WOBsAHClLolsH3Nf//2DVK5L8bpL3HvpEVaer6nxVnb948eLyKQHgCFgS3e0k1+85Pp7kmT3Hr05yc5JPVNXnk7wlydmDPkzV3We6e6u7t44dO3b5UwPABloS3UeTnKiqm6rqmiR3Jjn7fw9293PdfW1339jdNyZ5JMmp7j6/lokBYEMdGt3ufiHJPUkeSvK5JA929+NVdV9VnVr3gABwVFy9ZFF3n0tybt9977vE2rd+42MBwNHjilQAMER0AWCI6ALAENEFgCGiCwBDRBcAhoguAAwRXQAYIroAMER0AWCI6ALAENEFgCGiCwBDRBcAhoguAAwRXQAYIroAMER0AWCI6ALAENEFgCGiCwBDRBcAhoguAAwRXQAYIroAMER0AWCI6ALAENEFgCGiCwBDRBcAhoguAAwRXQAYIroAMER0AWCI6ALAENEFgCGiCwBDRBcAhoguAAwRXQAYIroAMER0AWCI6ALAENEFgCGiCwBDRBcAhoguAAwRXQAYIroAMGRRdKvq1qp6sqouVNW9Bzz+y1X1RFU9VlV/U1XfvfpRAWCzHRrdqroqyf1JbktyMsldVXVy37LPJNnq7h9I8udJ3r/qQQFg0y15pXtLkgvd/VR3P5/kgSR37F3Q3Q9395d3Dx9Jcny1YwLA5lsS3euSPL3neHv3vku5O8lfH/RAVZ2uqvNVdf7ixYvLpwSAI2BJdOuA+/rAhVXvTLKV5AMHPd7dZ7p7q7u3jh07tnxKADgCrl6wZjvJ9XuOjyd5Zv+iqnp7kl9L8qPd/dXVjAcAR8eSV7qPJjlRVTdV1TVJ7kxydu+CqnpTkg8lOdXdz65+TADYfIdGt7tfSHJPkoeSfC7Jg939eFXdV1Wndpd9IMmrkny8qv6xqs5e4ukA4GVryenldPe5JOf23fe+PbffvuK5AODIcUUqABgiugAwRHQBYIjoAsAQ0QWAIaILAENEFwCGiC4ADBFdABgiugAwRHQBYIjoAsAQ0QWAIaILAENEFwCGiC4ADBFdABgiugAwRHQBYIjoAsAQ0QWAIaILAENEFwCGiC4ADBFdABgiugAwRHQBYIjoAsAQ0QWAIaILAENEFwCGiC4ADBFdABgiugAwRHQBYIjoAsAQ0QWAIaILAENEFwCGiC4ADBFdABgiugAwRHQBYIjoAsAQ0QWAIaILAENEFwCGiC4ADFkU3aq6taqerKoLVXXvAY9/a1V9bPfxT1XVjaseFAA23aHRraqrktyf5LYkJ5PcVVUn9y27O8mXuvt7k/xukt9e9aAAsOmWvNK9JcmF7n6qu59P8kCSO/atuSPJH+/e/vMkb6uqWt2YALD5lkT3uiRP7zne3r3vwDXd/UKS55J81yoGBICj4uoFaw56xdqXsSZVdTrJ6d3Dr1bVZxf8fF6aa5P8+5Ue4oiyt+thX9fH3q7H913uNy6J7naS6/ccH0/yzCXWbFfV1Ulek+Q/9j9Rd59JciZJqup8d29dztBcmn1dH3u7HvZ1feztelTV+cv93iWnlx9NcqKqbqqqa5LcmeTsvjVnk/zM7u13JPnb7v66V7oA8HJ26Cvd7n6hqu5J8lCSq5J8uLsfr6r7kpzv7rNJ/ijJn1bVhey8wr1znUMDwCZacno53X0uybl9971vz+2vJPnJl/izz7zE9SxjX9fH3q6HfV0fe7sel72v5SwwAMxwGUgAGLL26LqE5Hos2NdfrqonquqxqvqbqvruKzHnJjpsb/ese0dVdVX5dOgCS/a1qn5q9/f28ar6s+kZN9GCvwU3VNXDVfWZ3b8Ht1+JOTdNVX24qp691D9trR2/t7vvj1XVmxc9cXev7Ss7H7z65yTfk+SaJP+U5OS+NT+X5IO7t+9M8rF1znQUvhbu648l+fbd2z9rX1e3t7vrXp3kk0keSbJ1pef+Zv9a+Dt7Islnknzn7vHrrvTc3+xfC/f1TJKf3b19Msnnr/Tcm/CV5EeSvDnJZy/x+O1J/jo716l4S5JPLXnedb/SdQnJ9Th0X7v74e7+8u7hI9n599UcbsnvbJL8VpL3J/nK5HAbbMm+vifJ/d39pSTp7meHZ9xES/a1k3zH7u3X5Ouvs8ABuvuTOeB6E3vckeRPescjSV5bVa8/7HnXHV2XkFyPJfu6193Z+S8yDnfo3lbVm5Jc391/NTnYhlvyO/uGJG+oqr+vqkeq6tax6TbXkn39zSTvrKrt7PwrlF+YGe3Ie6l/h5Ms/CdD34CVXUKSr7F4z6rqnUm2kvzoWic6Ol50b6vqFdn5P2m9a2qgI2LJ7+zV2TnF/NbsnJn5u6q6ubv/c82zbbIl+3pXko909+9U1Q9l55oKN3f3/6x/vCPtstq17le6L+USknmxS0jyNZbsa6rq7Ul+Lcmp7v7q0Gyb7rC9fXWSm5N8oqo+n533cs76MNWhlv4t+Mvu/q/u/pckT2Ynwlzakn29O8mDSdLd/5Dk27JzTWa+MYv+Du+37ui6hOR6HLqvu6dAP5Sd4HpvbLkX3dvufq67r+3uG7v7xuy8X36quy/7WqwvE0v+FvxFdj4AmKq6Njunm58anXLzLNnXLyR5W5JU1fdnJ7oXR6c8ms4m+endTzG/Jclz3f1vh33TWk8vt0tIrsXCff1Aklcl+fju59K+0N2nrtjQG2Lh3vISLdzXh5L8RFU9keS/k/xKd3/xyk39zW/hvr43yR9W1S9l5/Tnu7ywOVxVfTQ7b3Vcu/t++G8k+ZYk6e4PZuf98duTXEjy5STvXvS89h4AZrgiFQAMEV0AGCK6ADBEdAFgiOgCwBDRBYAhogsAQ0QXAIb8Lz1bmKyTNBysAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_barplot(results, test='num_genomes', metric='classification')\n",
    "plot_barplot(results, test='num_genomes', metric='abundance')\n",
    "\n",
    "results['num_genomes']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Number of read mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_barplot(results, test='num_reads_200genomes', metric='classification')\n",
    "plot_barplot(results, test='num_reads_200genomes', metric='abundance')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Number of reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_barplot(results, test='num_reads_800genomes', metric='classification')\n",
    "plot_barplot(results, test='num_reads_800genomes', metric='abundance')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/gpfs/summit/scratch/dral3008/BENCHMARKING_RESULTS/src/benchmarking\n"
     ]
    }
   ],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
